// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/forms/controls/RadioButtonGroup.js
"use strict";

import UIComponent from "../../../../base";
import { RadioButtonGroup } from "../../select-group";
import { unique } from "../../../../helpers/core/generators";
import {
    renderCriticalitySection,
    renderDualColumnView,
    renderErrorText,
    renderHelperText,
    renderTitle,
    renderSublabelContent,
    replaceChild
} from "../helpers/view";
import {
    generateValueChangePayload
} from "../helpers/props";

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------
const FormRadioButtonGroupEvents = {
    CHANGE: "FormRadioButtonGroup::change"
};

const namespace = "mpageui-FormRadioButtonGroup";

/**
 * The FormRadioButtonGroup class.
 * @class FormRadioButtonGroup
 */
class FormRadioButtonGroup extends UIComponent {
    /**
     * @inheritDoc
     */
    constructor(props, children) {
        super(props, children);
        this._state = {
            showErrorText: false,
            showHelperText: true,
            value: null
        };
        this._runRefreshItems = false;
    }

    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            criticalityHandler: null,
            completionHandler: null,
            display: "",
            errorHandler: null,
            formName: unique(namespace),
            hasInlineLabel: false,
            helperHandler: null,
            isDisabled: false,
            isRequired: false,
            showErrorText: false,
            showHelperText: true,
            showRequiredIndicator: true,
            sublabelContent: null,
            options: {},
            value: null,
            valueChangeEventName: FormRadioButtonGroupEvents.CHANGE
        };
    }

    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return {
            formName: () => { this._runRefreshItems = true; },
            isDisabled: () => { this._runRefreshItems = true; },
            showErrorText: (showErrorText) => { this._state.showErrorText = showErrorText; },
            showHelperText: (showHelperText) => { this._state.showHelperText = showHelperText; },
            options: (options) => {
                this.getChild("radioButtonGroup").setProps(options);
                this._runRefreshItems = true;
            },
            value: (value = null) => {
                this._state.value = value;
                this._runRefreshItems = true;
            },
            sublabelContent: (sublabelContent) => { replaceChild(this, "sublabelContent", sublabelContent); }
        };
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                radioButtonGroup: new RadioButtonGroup(Object.assign({ isDisabled: null }, this.getProp("options")))
            }
        ];
    }

    /**
     * @inheritDoc
     */
    afterCreate() {
        this.on(RadioButtonGroup.EVENTS.SELECTION_CHANGE, (ctx, payload) => {
            this.stopPropagation(RadioButtonGroup.EVENTS.SELECTION_CHANGE);

            const value = payload.selected.ids[ 0 ] || null;

            this._state = {
                showErrorText: true,
                showHelperText: true,
                value
            };

            this.emit(this.getProp("valueChangeEventName"), this,
                generateValueChangePayload(this.getProps(), value)
            );

            this.update();
        });
    }

    /**
     * refreshes the checkbox items
     * @returns {undefined} undefined
     * @private
     */
    _refreshItems() {
        const value = this._state.value;
        const radioButtonGroup = this.getChild("radioButtonGroup");
        const items = radioButtonGroup.getProp("items");
        const isDisabled = this.getProp("isDisabled");
        const formName = this.getProp("formName");

        radioButtonGroup.setProp("items", items.map((item, i) =>
            Object.assign({}, item,
                {
                    isSelected: value === String(item.id || i),
                    formName
                },
                isDisabled !== null ? { isDisabled } : null
            )
        ));
    }

    /**
     * @inheritDoc
     */
    beforeRender() {
        if (this._runRefreshItems) {
            this._refreshItems();
            this._runRefreshItems = false;
        }
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mappedChildren) {
        const state = this._state;
        return el(
            "div",
            {
                class: namespace
            },
            [
                renderDualColumnView(el, props.hasInlineLabel,
                    [
                        renderTitle(el, props),
                        renderSublabelContent(el, props, mappedChildren)
                    ],
                    [
                        mappedChildren.radioButtonGroup.render(),
                        renderErrorText(el, props, state),
                        renderCriticalitySection(el, props, state),
                        renderHelperText(el, props, state)
                    ]
                )
            ]
        );
    }
}

FormRadioButtonGroup.EVENTS = FormRadioButtonGroupEvents;

export default FormRadioButtonGroup;
