// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/forms/controls/TextBox.js
"use strict";

import UIComponent from "../../../../base";
import TextBox from "../../../atomic/textbox";
import { unique } from "../../../../helpers/core/generators";
import { forwardProp } from "../../../../helpers/props/props";
import {
    renderCriticalitySection,
    renderDualColumnView,
    renderErrorText,
    renderHelperText,
    renderTitle
} from "../helpers/view";
import {
    generateValueChangePayload
} from "../helpers/props";
import {
    characterCountHandler,
    generatePatternMatchErrorHandler,
    maxLengthHandler
} from "../helpers/handlers/text";

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------
const namespace = "mpageui-FormTextBox";

/**
 * The FormTextBox class.
 * @class FormTextBox
 */
class FormTextBox extends UIComponent {
    /**
     * @inheritDoc
     */
    constructor(props, children) {
        super(props, children);
        this._state = {
            showErrorText: false,
            showHelperText: true,
            value: ""
        };
    }

    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            blurEventName: FormTextBox.EVENTS.BLUR,
            criticalityHandler: null,
            completionHandler: (value) => value !== null && value !== "",
            display: "",
            errorHandler: FormTextBox.ERROR_HANDLERS.maxLen,
            focusEventName: FormTextBox.EVENTS.FOCUS,
            formName: unique(namespace),
            hasInlineLabel: false,
            helperHandler: FormTextBox.HELPER_HANDLERS.charCount,
            isDisabled: false,
            isRequired: false,
            maxLength: null,
            showErrorText: false,
            showHelperText: false,
            showRequiredIndicator: true,
            options: {},
            value: "",
            valueChangeEventName: FormTextBox.EVENTS.CHANGE
        };
    }

    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return {
            isDisabled: forwardProp(this, "isDisabled", "textBox"),
            formName: forwardProp(this, "name", "textBox"),
            showErrorText: (showErrorText) => { this._state.showErrorText = showErrorText; },
            showHelperText: (showHelperText) => { this._state.showHelperText = showHelperText; },
            options: (options) => {
                this.getChild("textBox").setProps(options);
            },
            value: (value = "") => {
                this._state.value = value;
                this.getChild("textBox").setProp("value", value === null ? "" : value);
            }
        };
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                textBox: new TextBox(this.getProp("options"))
            }
        ];
    }

    /**
     * @inheritDoc
     */
    afterCreate() {
        this.on(TextBox.EVENTS.INPUT, (ctx, value) => {
            this.stopPropagation(TextBox.EVENTS.INPUT);

            this._state.showHelperText = true;
            this._state.value = value;

            this.emit(this.getProp("valueChangeEventName"), this,
                generateValueChangePayload(this.getProps(), value)
            );

            this.update();
        });

        this.on(TextBox.EVENTS.BLUR, () => {
            this.stopPropagation(TextBox.EVENTS.BLUR);

            if (this._state.value !== "") {
                this._state.showHelperText = true;
                this._state.showErrorText = true;
            }

            this.emit(this.getProp("blurEventName"), this);
            this.update();
        });

        this.on(TextBox.EVENTS.FOCUS, () => {
            this.stopPropagation(TextBox.EVENTS.FOCUS);

            this._state.showHelperText = true;

            this.emit(this.getProp("focusEventName"), this);
            this.update();
        });
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mappedChildren) {
        const state = this._state;
        return el(
            "div",
            {
                class: namespace
            },
            [
                renderDualColumnView(el, props.hasInlineLabel,
                    [
                        renderTitle(el, props)
                    ],
                    [
                        mappedChildren.textBox.render(),
                        renderErrorText(el, props, state),
                        renderCriticalitySection(el, props, state),
                        renderHelperText(el, props, state)
                    ]
                )
            ]
        );
    }
}

FormTextBox.EVENTS = {
    BLUR: "FormTextBox::blur",
    CHANGE: "FormTextBox::change",
    FOCUS: "FormTextBox::focus"
};
FormTextBox.HELPER_HANDLERS = {
    charCount: characterCountHandler
};
FormTextBox.ERROR_HANDLERS = {
    maxLen: maxLengthHandler,
    generatePatternMatchHandler: generatePatternMatchErrorHandler
};

export default FormTextBox;
