// MPageFusion/testing/src/pages/controls/composite/form/controls/FormTextBox.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _FormControl = _interopRequireDefault(require("../FormControl.page"));

var _TextBox = _interopRequireDefault(require("../../../atomic/textbox/TextBox.page"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-FormTextBox";
/**
 * The FormTextBoxPage class
 * @class module:MPageFusion/testing.composite.form.FormTextBoxPage
 *
 * @example
 * const formTextBox = new FormTextBoxPage(driver);
 * assert(formTextBox.display).equals("Sample TextBox Label");
 */

var FormTextBoxPage =
/*#__PURE__*/
function (_FormControlPage) {
  _inheritsLoose(FormTextBoxPage, _FormControlPage);

  function FormTextBoxPage(driver, options) {
    var _this;

    _this = _FormControlPage.call(this, driver, options) || this;
    _this._textbox = new _TextBox["default"](driver, {
      selectorList: _this.selectorList
    });
    return _this;
  }

  var _proto = FormTextBoxPage.prototype;

  /**
   * checks the focused element
   * @memberOf module:MPageFusion/testing.composite.form.FormTextBoxPage
   * @returns {Promise} resolves when the action is completed
   */
  _proto.focus = function focus() {
    return this._textbox.focus();
  }
  /**
   * sendkeys to the textbox
   * @memberOf module:MPageFusion/testing.composite.form.FormTextBoxPage
   * @param {String} text - the input text
   * @returns {Promise} resolves when the action is completed
   */
  ;

  _proto.enterText = function enterText(text) {
    return this._textbox.enterText(text);
  }
  /**
   * clears text from the textbox
   * @memberOf module:MPageFusion/testing.composite.form.FormTextBoxPage
   * @returns {Promise} resolves when the action is completed
   */
  ;

  _proto.clearText = function clearText() {
    return this._textbox.clearText();
  };

  _createClass(FormTextBoxPage, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * the value in the textbox
     * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
     * @returns {Promise<String>} display text
     */

  }, {
    key: "value",
    get: function get() {
      return this._textbox.value;
    }
    /**
     * the placeholder
     * @memberOf module:MPageFusion/testing.composite.form.FormTextBoxPage
     * @returns {Promise<String>} display placeholder
     */

  }, {
    key: "placeholder",
    get: function get() {
      return this._textbox.placeholder;
    }
  }]);

  return FormTextBoxPage;
}(_FormControl["default"]);

exports["default"] = FormTextBoxPage;
