// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/tag-list/search/const.js
const EVENTS = {
    INPUT: "TagListSearch::input",
    INPUT_FOCUS: "TagListSearch::inputFocus",
    INPUT_BLUR: "TagListSearch::inputBlur",
    REQUEST_CONFIRM: "TagListSearch::requestConfirm",
    REQUEST_DELETE: "TagListSearch::requestDelete",
    REQUEST_SUGGESTIONS: "TagListSearch::requestSuggestions",
    REQUEST_CLEAR_SUGGESTIONS: "TagListSearch::requestClearSuggestions"
};

const NAMESPACE = "mpageui-TagListSearch";

/**
 * Enum of private events.
 * @type {{REQUEST_SHOW_ALL: string, REQUEST_DELETE: string}}
 */
const PRIVATE_EVENTS = {
    REQUEST_SHOW_ALL: "__TagListSearch::TagList::requestShowAll__",
    REQUEST_DELETE: "__TagListSearch::TagList::requestDelete__",
    RESULT_CLICK: "__TagListSearch::SearchList::resultClick__"
};

/**
 * Enum of input types.
 * @type {{TYPE: string, CLICK: string, ARROW_UP: string, ARROW_DOWN: string, ESCAPE: string}}
 */
const INPUT_TYPES = {
    TYPE: "type",
    CLICK: "click",
    ARROW_UP: "up",
    ARROW_DOWN: "down",
    ESCAPE: "escape"
};

/**
 * Enum of confirm types
 * @type {{ENTER: string, CLICK: string}}
 */
const CONFIRM_TYPES = {
    ENTER: "enter",
    CLICK: "click"
};

export {
    CONFIRM_TYPES,
    INPUT_TYPES,
    EVENTS,
    NAMESPACE,
    PRIVATE_EVENTS
};
