// MPageFusion/testing/src/pages/controls/atomic/table/TableSubsection.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _seleniumWebdriver = require("selenium-webdriver");

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _TableRow = _interopRequireDefault(require("./TableRow.page"));

var _List = _interopRequireDefault(require("../../../helpers/List"));

var _SvgIcon = _interopRequireDefault(require("../icon/svg/SvgIcon.page"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-Table-sectionContainer";
var HEADER = ".mpageui-Table-sectionHeader";
var ROW_CONTAINER = ".mpageui-Table-rowContainer";
/**
 * The TableSubsectionPage class
 * @alias module:MPageFusion/testing.atomic.table.TablePage.subsection
 * @class TableSubsectionPage
 * @private
 */

var TableSubsection =
/*#__PURE__*/
function (_UIComponentPage) {
  _inheritsLoose(TableSubsection, _UIComponentPage);

  function TableSubsection() {
    return _UIComponentPage.apply(this, arguments) || this;
  }

  var _proto = TableSubsection.prototype;

  /**
   * clicks the subsection header
   * @memberOf module:MPageFusion/testing.atomic.table.TablePage.subsection
   * @returns {Promise} - resolves when the action has completed
   */
  _proto.click = function click() {
    return this.getRootElement().then(function (element) {
      return element.findElement(_seleniumWebdriver.By.css("" + HEADER));
    }).then(function (element) {
      return element.click();
    });
  }
  /**
   * expands the subsection
   * @memberOf module:MPageFusion/testing.atomic.table.TablePage.subsection
   * @returns {Promise} - resolves when the action has completed
   */
  ;

  _proto.expand = function expand() {
    return this.getRootElement().then(function (element) {
      return element.findElement(_seleniumWebdriver.By.css("" + HEADER));
    }).then(function (element) {
      return new Promise(function (resolve) {
        element.getAttribute("class").then(function (classNames) {
          if (!classNames.includes("closed")) {
            throw new Error("Table subsection not collapsed");
          }

          resolve(element);
        });
      });
    }).then(function (element) {
      return element.click();
    });
  }
  /**
   * collapses the subsection
   * @memberOf module:MPageFusion/testing.atomic.table.TablePage.subsection
   * @returns {Promise} - resolves when the action has completed
   */
  ;

  _proto.collapse = function collapse() {
    return this.getRootElement().then(function (element) {
      return element.findElement(_seleniumWebdriver.By.css("" + HEADER));
    }).then(function (element) {
      return new Promise(function (resolve) {
        element.getAttribute("class").then(function (classNames) {
          if (classNames.includes("closed")) {
            throw new Error("Table subsection not expanded");
          }

          resolve(element);
        });
      });
    }).then(function (element) {
      return element.click();
    });
  };

  _createClass(TableSubsection, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * gets the subsection header display
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage.subsection
     * @returns {Promise<string>} headerDisplay
     */

  }, {
    key: "headerDisplay",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.findElement(_seleniumWebdriver.By.css(HEADER + " > span"));
      }).then(function (element) {
        return element.getText();
      });
    }
    /**
     * gets the subsection header icon
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage.subsection
     * @returns {SvgIconPage} headerIcon
     */

  }, {
    key: "headerIcon",
    get: function get() {
      return new _SvgIcon["default"](this.driver, {
        index: 1,
        selectorList: this.selectorList
      });
    }
    /**
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage.subsection
     * @returns {List<module:MPageFusion/testing.atomic.table.TablePage.row>} row
     */

  }, {
    key: "row",
    get: function get() {
      return (0, _List["default"])({
        list: this,
        itemPage: _TableRow["default"],
        selectorList: this.selectorList,
        containerSelector: ROW_CONTAINER
      });
    }
    /**
     * the subsection data
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage.subsection
     * @returns {Promise<object>} data
     */

  }, {
    key: "data",
    get: function get() {
      var _this = this;

      return this.get("isExpanded").then(function (isExpanded) {
        if (!isExpanded) {
          throw new Error("Cannot read data while there are collapsed subsections.");
        }

        return Promise.all([_this.headerDisplay, _this.row.all("data")]).then(function (_ref) {
          var _ref2 = _slicedToArray(_ref, 2),
              headerDisplay = _ref2[0],
              rows = _ref2[1];

          return {
            headerDisplay: headerDisplay,
            rows: rows
          };
        });
      });
    }
    /**
     * whether or not the subsection is expanded
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage.subsection
     * @returns {Promise<boolean>} isExpanded
     */

  }, {
    key: "isExpanded",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.findElement(_seleniumWebdriver.By.css("" + HEADER));
      }).then(function (element) {
        return element.getAttribute("class");
      }).then(function (classNames) {
        return !classNames.includes("closed");
      });
    }
  }]);

  return TableSubsection;
}(_UIComponent["default"]);

exports["default"] = TableSubsection;
