// MPageFusion/testing/src/pages/controls/atomic/table/Table.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _seleniumWebdriver = require("selenium-webdriver");

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _List = _interopRequireDefault(require("../../../helpers/List"));

var _TableSubsection = _interopRequireDefault(require("./TableSubsection.page"));

var _TableRow = _interopRequireDefault(require("./TableRow.page"));

var _TableColumn = _interopRequireDefault(require("./TableColumn.page"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-Table";
var BODY = ".mpageui-Table-body";
var ROW_CONTAINER = ".mpageui-Table-rowContainer";
var ROW = ".mpageui-Table-row";
var HEADER = ".mpageui-Table-header";
var SUBSECTION_HEADER = ".mpageui-Table-sectionHeader";
/**
 * The TablePage class
 * @class module:MPageFusion/testing.atomic.table.TablePage
 *
 * @example
 * const table = new TablePage(driver);
 * table.subsection(0).row(4).click();
 */

var TablePage =
/*#__PURE__*/
function (_UIComponentPage) {
  _inheritsLoose(TablePage, _UIComponentPage);

  function TablePage() {
    return _UIComponentPage.apply(this, arguments) || this;
  }

  _createClass(TablePage, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage
     * @returns {List<module:MPageFusion/testing.atomic.table.TablePage.subsection>} subsection
     */

  }, {
    key: "subsection",
    get: function get() {
      return (0, _List["default"])({
        list: this,
        itemPage: _TableSubsection["default"],
        selectorList: this.selectorList,
        containerSelector: BODY
      });
    }
    /**
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage
     * @returns {List<module:MPageFusion/testing.atomic.table.TablePage.row>} row
     */

  }, {
    key: "row",
    get: function get() {
      return (0, _List["default"])({
        list: this,
        itemPage: _TableRow["default"],
        selectorList: this.selectorList,
        containerSelector: ROW_CONTAINER
      });
    }
    /**
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage
     * @returns {List<module:MPageFusion/testing.atomic.table.TablePage.column>} column
     */

  }, {
    key: "column",
    get: function get() {
      return (0, _List["default"])({
        list: this,
        itemPage: _TableColumn["default"],
        selectorList: this.selectorList,
        containerSelector: HEADER + " " + ROW
      });
    }
    /**
     * the table data
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage
     * @returns {Promise<object>} data
     */

  }, {
    key: "data",
    get: function get() {
      var _this = this;

      return this.subsection(0).getRootElement() // table includes a subsection element even if not using subsections
      .then(function (element) {
        return (// check to see if subsections are used
          element.findElements(_seleniumWebdriver.By.css(SUBSECTION_HEADER)).then(function (headers) {
            return headers.length ? Promise.all([_this.subsection.all("data"), _this.column.all("data")]).then(function (_ref) {
              var _ref2 = _slicedToArray(_ref, 2),
                  subsections = _ref2[0],
                  columns = _ref2[1];

              return {
                subsections: subsections,
                columns: columns
              };
            }) : Promise.all([_this.row.all("data"), _this.column.all("data")]).then(function (_ref3) {
              var _ref4 = _slicedToArray(_ref3, 2),
                  rows = _ref4[0],
                  columns = _ref4[1];

              return {
                rows: rows,
                columns: columns
              };
            });
          })
        );
      });
    }
  }]);

  return TablePage;
}(_UIComponent["default"]);

exports["default"] = TablePage;
