// MPageFusion/testing/src/pages/controls/atomic/table/TableColumn.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-Table-row-cell";
var SIZES = {
  ONE: "1",
  TWO: "2",
  THREE: "3",
  FOUR: "4",
  FIVE: "5"
};
/**
 * The TableColumnPage class
 * @alias module:MPageFusion/testing.atomic.table.TablePage.column
 * @class TableColumnPage
 * @private
 */

var TableColumn =
/*#__PURE__*/
function (_UIComponentPage) {
  _inheritsLoose(TableColumn, _UIComponentPage);

  function TableColumn() {
    return _UIComponentPage.apply(this, arguments) || this;
  }

  var _proto = TableColumn.prototype;

  /**
   * clicks the column header
   * @memberOf module:MPageFusion/testing.atomic.table.TablePage.column
   * @returns {Promise} - resolves when the action has completed
   */
  _proto.sort = function sort() {
    return this.getRootElement().then(function (element) {
      return element.click();
    });
  }
  /**
   * gets the column size
   * @memberOf module:MPageFusion/testing.atomic.table.TablePage.column
   * @returns {Promise<string>} size
   */
  ;

  _createClass(TableColumn, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * the column header display
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage.column
     * @returns {Promise<string>} display
     */

  }, {
    key: "display",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.getText();
      });
    }
    /**
     * the column data
     * @memberOf module:MPageFusion/testing.atomic.table.TablePage.column
     * @returns {Promise<object>} data
     */

  }, {
    key: "data",
    get: function get() {
      return this.get("display").then(function (display) {
        return {
          display: display
        };
      });
    }
  }, {
    key: "sizeSpan",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.getAttribute("class");
      }).then(function (classNames) {
        return Object.values(SIZES).find(function (size) {
          return classNames.includes("mpageui-Table-row-cell--size" + size);
        }) || "";
      });
    }
  }]);

  return TableColumn;
}(_UIComponent["default"]);

exports["default"] = TableColumn;
