// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/search/pane/SearchPane.js
"use strict";
import UIComponent from "../../../../base/UIComponent";
import SearchList from "../list/SearchList";
import { adjustPaneHeight } from "../manager/helpers";
import { getLoadingSpinner } from "../helpers";
import { emitUsingProp, forwardProp } from "../../../../helpers/props/props";
const NAMESPACE = "mpageui-SearchPane";
const EVENTS = {
    OVERLAY_CLICK: "SearchPane::overlayClick"
};

/**
 * Results floating pane that loads in screen sizes MD up and
 * has the results list from suggestions.Shows up when the throttle time expires or on search submit
 */
class SearchPane extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            isLoading: false,
            isVisible: false,
            overlayClickEventName: EVENTS.OVERLAY_CLICK,
            results: [],
            searchValue: "",
            selectedIndex: 0,
            shouldTruncate: true
        };
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                searchList: new SearchList({
                    preventMouseDownDefault: true,
                    results: [],
                    shouldAutoScroll: true
                })
            }
        ];
    }

    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return {
            selectedIndex: forwardProp(this, "selectedIndex", "searchList"),
            results: forwardProp(this, "results", "searchList"),
            searchValue: forwardProp(this, "searchValue", "searchList"),
            shouldTruncate: forwardProp(this, "shouldTruncate", "searchList")
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mapped) {
        return props.isVisible ? [
            el(
                "div",
                {
                    class: `${NAMESPACE}-overlay`,
                    onclick: emitUsingProp("overlayClickEventName", this, this)
                }
            ),
            [
                el(
                    "div",
                    {
                        class: NAMESPACE,
                        oncreate: ({ dom }) => {
                            adjustPaneHeight(dom);
                        },
                        onupdate: ({ dom }) => {
                            adjustPaneHeight(dom);
                        }
                    },
                    props.isLoading ? (getLoadingSpinner().render()) : mapped.searchList.render()
                )
            ]
        ] : null;
    }
}

SearchPane.EVENTS = EVENTS;

export default SearchPane;
