// gaia-plugin-livedev/src/ui/gaia-plugin-livedev/driver-config/program-params/parseUnifiedOrgDriver.js
/**
 * Based on the old-school-cool param parser, takes an Organizer MPage View source string and generates a URL param.
 * @param {String} sourceString - Page source.
 * @returns {Object} Parsed params.
 */
const parse = (sourceString) => {
    var matchArr = [];
    const paramObj = {
        mine: "^MINE^",
        userId: "0.0",
        positionCd: "0.0",
        devLoc: '""',
        application: "^powerchart.exe^",
        staticContent: '^$STATIC_CONTENT$|MPAGES_xx|MPAGES_WORKLISTS_xx^',
        categoryMean: "",
        viewpointName: "",
        inBrowser: 73
    };

    matchArr = /"PRSNL_ID":[0-9]+\./.exec(sourceString);
    if (matchArr && matchArr.length) {
        paramObj.userId = matchArr[ 0 ].replace('"PRSNL_ID":', "").replace(".", ".0");
    }

    matchArr = /"POSITION_CD":[0-9]+\./.exec(sourceString);
    if (matchArr && matchArr.length) {
        paramObj.positionCd = matchArr[ 0 ].replace('"POSITION_CD":', "").replace(".", ".0");
    }

    matchArr = /"CATEGORY_MEAN":".*","LOCALE_ID/.exec(sourceString);
    if (matchArr && matchArr.length) {
        paramObj.categoryMean = matchArr[ 0 ].replace('"CATEGORY_MEAN":"', "^").replace('","LOCALE_ID', '^');
    }

    matchArr = /"VIEWPOINT_NAME_KEY":".*","ACTIVE_VIEW_CAT_MEAN"/.exec(sourceString);
    if (matchArr && matchArr.length) {
        paramObj.viewpointName = matchArr[ 0 ].replace('"VIEWPOINT_NAME_KEY":"', "^").replace('","ACTIVE_VIEW_CAT_MEAN"', '^');
    }

    return paramObj;
};

/**
 * Given the source and config, generates a parameterized URL.
 * @param {String} source - The MPage source string.
 * @param {Object} config - Config passed from the livedev plugin.
 * @returns {String} A parameterized URL.
 */
const generateParams = ({ source, config }) => {
    const args = parse(source);
    return [
        args.mine,
        args.userId,
        args.positionCd,
        args.application,
        args.staticContent,
        args.viewpointName,
        args.inBrowser
    ].join(",")
};

export default generateParams;
