// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/button/AsyncButton.js
"use strict";

import UIComponent from "../../../base";
import Button from "../../atomic/button";
import {
    forwardProps,
    forwardProp
} from "../../../helpers/props/props";

/**
 * Enumeration of events emitted by the AsyncButton.
 * @type {{CLICK: string}}
 */
const EVENTS = {
    CLICK: "AsyncButton::click"
};

/**
 * events private to the AsyncButton.
 * @type {{CLICK: string}}
 */
const PRIVATE_EVENTS = {
    CLICK: "__AsyncButton::click__"
};

/**
 * The AsyncButton class.
 * @class AsyncButton
 */
export default class AsyncButton extends UIComponent {

    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            isProcessing: false,
            clickEventName: EVENTS.CLICK,
            classNames: "",
            display: "",
            displayPosition: "after",
            icon: null,
            isSelected: false,
            label: "",
            size: "",
            type: "",
            responsiveIconSize: null
        };
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                button: new Button({
                    clickEventName: PRIVATE_EVENTS.CLICK
                }, [])
            }
        ];
    }

    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return Object.assign(
            forwardProps(this, "button", [
                "classNames",
                "display",
                "displayPosition",
                "icon",
                "isSelected",
                "label",
                "size",
                "type",
                "responsiveIconSize",
                "title"
            ]),
            {
                isProcessing: forwardProp(this, "disabled", "button")
            }
        );
    }

    /**
     * @inheritDoc
     */
    afterCreate() {
        this.redirectEventUsingProp(PRIVATE_EVENTS.CLICK, "clickEventName", this);
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mChildren) {
        return this.renderChildren();
    }
}
AsyncButton.EVENTS = EVENTS;
