// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/search/popup/FilterTab.js
"use strict";
import UIComponent from "../../../../base/UIComponent";
import Label from "../../../atomic/label";
import Expand from "../../../atomic/icon/svg/category/navigation/ExpandMore";
import { PRIVATE_EVENTS } from "./const";

const NAMESPACE = "mpageui-FilterTab";

/**
 * The FilterTab class renders a div containing
 * the selected value in the scopedSearch Control.
 * The FilterTab class is rendered
 * when the screen size is MD down from the Search Manager
 */
class FilterTab extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            classNames: "",
            value: ""
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mChildren) {
        return el(
            "div",
            {
                class: `${NAMESPACE}--wrapper`,
                onclick: (event) => this.emit(PRIVATE_EVENTS.FILTER_TAB_CLICK, event)
            },
            [
                new Label({
                    classNames: props.classNames,
                    value: props.value
                }).render(),
                new Expand().render()
            ]
        );
    }
}

export default FilterTab;
