// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/search/popup/SelectLabel.js
"use strict";
import UIComponent from "../../../../base/UIComponent";
import Label from "../../../atomic/label";
import { PRIVATE_EVENTS } from "./const";

const NAMESPACE = "mpageui-SelectLabel";

/**
 * SelectLabel contains a label with a click event.
 * SelectLabel class is rendered
 * when the screen size is MD down.
 */
class SelectLabel extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            classNames: "",
            value: ""
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mChildren) {
        return el(
            "div",
            {
                class: NAMESPACE,
                onclick: (event) => this.emit(PRIVATE_EVENTS.SELECT_LABEL_BACK, this.getProp("value"), event)
            },
            [
                new Label({
                    classNames: props.classNames,
                    value: props.value
                }).render()
            ]
        );
    }
}

export default SelectLabel;
