// gaia-plugin-whats-up-doc/src/gaiaPluginWhatsUpDoc.js
const api = require("./commands/api");
const config = require("./commands/config");

module.exports = {
    id: "whats-up-doc",
    commands: {
        toMarkdown: {
            description: "Generate a markdown string of the API documentation",
            handler: api.toMarkdownString
        },
        toFile: {
            description: "Generate a markdown file of the API documentation",
            handler: api.toMarkdownFile
        },
        config: {
            description: "Update the WhatsUpDoc configuration.",
            args: {
                markdownOutputLocation: {
                    description: "The name of the file to which a markdown file can be exported. Ex: 'API.md'"
                },
                files: {
                    description: "A globbed file pattern. Ex: 'src/main/**/*.js'"
                }
            },
            handler: config.set
        },
        getGuide: {
            description: "Retrieve all the guide markdown files.",
            handler: api.getGuide
        }
    },
    ui: "dist/ui.js"
};
