// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/popup-list/PopupListContents.js
import UIComponent from "../../../base/UIComponent";
import { emitUsingProp } from "../../../helpers/props/props";
import { svg } from "../../atomic/icon";
import classNames from "classnames";
import { isSelectedClassName } from "../../../helpers/const/classNames";

/**
 * The contents of the PopupList, specified as an unordered list.
 * The list items have radio buttons in order to be aria compliant.
 */
export default class PopupListContents extends UIComponent {
    constructor(props, children) {
        super(props, children);
    }

    /**
     * @inheritdoc
     */
    initialProps() {
        return {
            items: [],
            itemSelectEventName: "PopupList::itemSelect",
            classNames: "",
            defaultClassNames: "mpageui-PopupList",
            isSelectionVisible: true
        };
    }

    /**
     * @inheritdoc
     */
    view(el, props) {
        return el("ul",
            {
                class: classNames({
                    [ props.defaultClassNames ]: true,
                    [ props.classNames ]: props.classNames.length > 0,
                    "mpageui-PopupList--withselection": props.isSelectionVisible
                })
            },
            props.items.map(
                (item) =>
                    el("li",
                        {
                            onclick: emitUsingProp("itemSelectEventName", this, item),
                            class: classNames({
                                [ item.classNames ]: item.classNames && item.classNames.length > 0,
                                [ isSelectedClassName ]: item.isSelected
                            })
                        },
                        [
                            props.isSelectionVisible && item.isSelected ? (new svg.content.Checkmark()).render() : null,
                            props.isSelectionVisible ? el("input", { type: "radio", checked: item.isSelected }) : null,
                            item.display
                        ])
            ));
    }
}
