// MPageFusion/build/webpack/config/dev.config.js
"use strict";

const path = require("path");
const webpack = require("webpack");
const withLocalI18n = require("../helpers/i18n").withLocalI18n;
const { baseConfig: babelBaseConfig } = require("../helpers/babel");

module.exports = withLocalI18n({
    devtool: "inline-source-map",
    entry: [
        "webpack-dev-server/client?http://localhost:1337/",
        "./dev/app.js"
    ],
    output: {
        path: path.join(process.cwd(), "dist"),
        filename: "bundle.js",
        publicPath: "/dist/"
    },
    resolve: {
        modules: [
            process.cwd(),
            "node_modules"
        ]
    },
    module: {
        rules: [
            {
                test: /\.(js|jsx)$/,
                include: [
                    path.join(__dirname, "../../.." , "src"),
                    path.join(__dirname, "../../..", "dev")
                ],
                use: [
                    babelBaseConfig("DEV")
                ]
            },
            {
                test: /\.(png|svg)$/,
                loader: "url-loader"
            },
            {
                test: /\.(css|less)$/,
                loader: "style-loader!css-loader!less-loader"
            }
        ]
    },
    plugins: [
        new webpack.optimize.ModuleConcatenationPlugin({})
    ]
});
