// MPageFusion/build/webpack/config/webpack.mpage.config.js
"use strict";

const { baseConfig, jsLoaderConfig } = require("./webpack.common");
const { baseConfig: babelBaseConfig } = require("../helpers/babel");
const WrapperPlugin = require("wrapper-webpack-plugin");
const merge = require("webpack-merge");
const version = require("../../../package.json").version;
const { withGlobalI18n } = require("../helpers/i18n");

const getDevConfig = (env) =>
    env && env.TYPE === "DEV" ?
        {
            devtool: "eval-source-map"
        } :
        null;

const config = (env) => withGlobalI18n(merge(
    baseConfig(),
    {
        module: {
            rules: [
                jsLoaderConfig("MPAGE", version, babelBaseConfig)
            ]
        }
    },
    {
        plugins: [
            new WrapperPlugin({
                header: function(fileName){
                    if(fileName.indexOf("mpage-fusion.js") > -1){
                        return '(function () { "use strict";\n var FUSION_LOCALE = window.m_localeObjectName;\n  ';
                    }
                    return "";
                },
                footer: function(fileName){
                    if(fileName.indexOf("mpage-fusion.js") > -1){
                        return '\n if(typeof MPageUI === "undefined"){window.MPageUI = {}}; \n MPageUI.Fusion = window.MPageFusion;\n})();';
                    }
                    return "";
                }
            })
        ]
    },
    getDevConfig(env)
));

module.exports = config;
