// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/table/TableSectionContainer.js
/**
 * @fileoverview TableSectionContainer class
 * @author Gyandeep Singh
 */

"use strict";

import UIComponent from "../../../base";
import TableSectionHeader from "./TableSectionHeader";
import TableRowContainer from "./TableRowContainer";
import classNames from "classnames";

/**
 * Constructor for TableSectionContainer
 * @extends UIComponent
 */
export default class TableSectionContainer extends UIComponent {
    /**
     * @inheritdoc
     */
    initialProps() {
        return {
            defaultClassNames: "mpageui-Table-sectionContainer",
            classNames: "",
            rows: [],
            rowProps: {},
            display: "",
            sectionIcon: null,
            key: "",
            isBodyVisible: true,
            columns: [],
            index: 0,
            cardView: false
        };
    }

    /**
     * @inheritdoc
     */
    createChildren() {
        const children = [];
        const headerDisplay = this.getProp("display");
        const isBodyVisible = this.getProp("isBodyVisible");

        if (headerDisplay.length > 0) {
            children.push({
                header: new TableSectionHeader({
                    display: headerDisplay,
                    sectionIcon: this.getProp("sectionIcon"),
                    isVisible: isBodyVisible
                }, [])
            });
        }

        children.push({
            body: new TableRowContainer({
                rows: this.getProp("rows"),
                rowProps: this.getProp("rowProps"),
                isVisible: isBodyVisible,
                columns: this.getProp("columns"),
                index: this.getProp("index"),
                cardView: this.getProp("cardView")
            })
        });

        return children;
    }

    /**
     * @inheritdoc
     */
    view(el, props, children, mappedChildren) {
        return el("div",
            {
                class: classNames({
                    [ props.defaultClassNames ]: true,
                    [ props.classNames ]: props.classNames.length > 0
                })
            },
            this.renderChildren()
        );
    }
}
