// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/table/TableActionContent.js
/**
 * @fileoverview Wrapper for action cell content
 * @author Gyandeep Singh
 */

"use strict";

import UIComponent from "../../../base";

/**
 * Constructor for TableActionContent
 * Purpose of this class is that it will stop click events from propagating. This way for cells which
 * have action and whe user clicks on those actions the row selection event is prevented.
 * @extends UIComponent
 */
export default class TableActionContent extends UIComponent {
    /**
     * @inheritdoc
     */
    initialProps() {
        return {
            content: []
        };
    }

    /**
     * @inheritdoc
     */
    propChangeHandlers() {
        return {
            content: (content) => this.replaceAllChildren(content)
        };
    }

    /**
     * @inheritdoc
     */
    view(el) {
        return el(
            "div",
            {
                class: "mpageui-Table-row-cell-content",
                // reason for this is explained above under the class description
                /* istanbul ignore next : can be tested when we use real dom so that event bubble up */
                onclick: /* istanbul ignore next */ (event) => event.stopPropagation()
            },
            this.renderChildren()
        );
    }
}
