// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/sticky-container/StickyContainer.js
import UIComponent from "../../../base/UIComponent";
import { unique } from "../../../helpers/core/generators";

/**
 * A container that will stick to the top of the first scrollpane
 * in the parent chain.
 */
export default class StickyContainer extends UIComponent {
    constructor(props, children) {
        super(props, children);
        this._referenceElementId = unique("reference_element");
        this._contentElementId = unique("content_element");
    }

    /**
     * @inheritdoc
     */
    initialProps() {
        return {
            className: "mpageui-StickyContainer",
            referenceElementId: unique("reference_element"),
            contentElementId: unique("content_element"),
            scrollPaneSlot: "default"
        };
    }

    /**
     * @inheritdoc
     */
    view(el, props, children) {
        return el(
            "div",
            {
                id: props.referenceElementId,
                class: `${props.className}-reference`
            },
            el(
                "div",
                {
                    id: props.contentElementId,
                    class: props.className
                },
                this.renderChildren()
            )
        );
    }
}
