// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/viewpoint-tab-list/ViewpointTabLink.js
import UIComponent from "../../../base";
import { svg } from "../../atomic/icon";

const { navigation: { Close } } = svg;

const namespace = "mpageui-ViewpointTab";
const ViewpointTabLinkEvents = {
    CLOSE_CLICK: "ViewpointTabLink::closeClick"
};

/**
 * Class representing the contents of a Viewpoint Tab.
 * @class ViewpointTabLink
 */
class ViewpointTabLink extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            id: "",
            display: "",
            title: ""
        };
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            new Close({
                classNames: `${namespace}-closeIcon`
            })
        ];
    }

    /**
     * @inheritDoc
     */
    view(el, props, children) {
        return el(
            "a",
            {
                class: `${namespace}-link`,
                title: props.title
            },
            [
                el(
                    "span",
                    {
                        class: `${namespace}-display`
                    },
                    props.display
                ),
                el(
                    "a",
                    {
                        class: `${namespace}-close`,
                        role: "button",
                        draggable: false,
                        onclick: (event) => {
                            // Prevent triggering other events
                            event.stopPropagation();
                            this.emit(ViewpointTabLinkEvents.CLOSE_CLICK, props.id);
                        }
                    },
                    this.renderChildren()
                )
            ]
        );
    }
}

ViewpointTabLink.EVENTS = ViewpointTabLinkEvents;

export default ViewpointTabLink;
