// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/search/popup/SearchFilterHeader.js
"use strict";
import UIComponent from "../../../../base/UIComponent";

/**
 * The SearchFilterHeader class consist of 2 header objects:
 * - searchHeader:
 *      -- contains the Navigation Header for the search manager
 * - filterHeader:
 *      --  contains the popup instance of the filter tab for Scopedsearch control.
 * - hasFilterHeader
 *      -- boolean to decide whether to render the filterHeader for the ScopedSearch control
 *  This class is triggered when the screen size is MD down from the Search Manager
 */
class SearchFilterHeader extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            searchHeader: null,
            filterHeader: null,
            hasFilterHeader: false
        };
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                searchHeader: this.getProp("searchHeader")
            },
            {
                filterHeader: this.getProp("filterHeader")
            }
        ];
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mChildren) {
        return [
            mChildren.searchHeader.render(),
            props.hasFilterHeader ? el(
                "div",
                {
                    class: "mpageui-u-display-flex"
                },
                mChildren.filterHeader.render()) : null
        ];
    }
}

export default SearchFilterHeader;
