// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/select-group/ButtonGroup.js
"use strict";

import SelectGroup from "./SelectGroup";
import SelectButton from "../../atomic/button/SelectButton";

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------
const ButtonGroupEvents = {
    SELECTION_CHANGE: "ButtonGroup::selectionChange"
};

const namespace = "mpageui-ButtonGroup";

/**
 * SelectButton generator
 * @param {Object} props - SelectButton properties
 * @returns {SelectButton} - SelectButton component
 */
const generateSelectItem = (props) => new SelectButton(props, []);

/**
 * The ButtonGroup class.
 * @class ButtonGroup
 * @augments  SelectGroup
 */
export default class ButtonGroup extends SelectGroup {

    /**
     * @inheritDoc
     */
    initialProps() {
        return Object.assign(super.initialProps(), {
            defaultClassNames: namespace,
            selectionChangeEventName: ButtonGroupEvents.SELECTION_CHANGE,
            itemGenerator: generateSelectItem
        });
    }
}
