// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/panel/detail-panel/DetailPanelHeader.js
"use strict";

import UIComponent from "../../../../base/UIComponent";
import { NavigationHeader } from "../../../atomic/header";

const DETAIL_PANEL_HEADER_EVENTS = {
    DISMISS: "DetailPanelHeader::dismiss"
};

/**
 * The DetailPanelHeader class.
 * @class DetailPanelHeader
 */
class DetailPanelHeader extends UIComponent {
    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                contentHeader: new NavigationHeader({
                    classNames: "mpageui-DetailPanel-headerSection",
                    showDismiss: true
                })
            }
        ];
    }

    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            title: "",
            toolbarContent: null
        };
    }

    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return {
            toolbarContent: (toolbarContent) => {
                this.getChild("contentHeader").setProp("content", [ toolbarContent ]);
            },
            title: (title) => {
                this.getChild("contentHeader").setProp("title", title);
            }
        };
    }

    /**
     * @inheritDoc
     */
    afterCreate() {
        this.redirectEvent(NavigationHeader.EVENTS.DISMISS, DetailPanelHeader.EVENTS.DISMISS, this);
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mappedChildren) {
        return mappedChildren.contentHeader.render();
    }
}

DetailPanelHeader.EVENTS = DETAIL_PANEL_HEADER_EVENTS;

export default DetailPanelHeader;
