// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/panel/detail-panel/DetailPanelBanners.js
"use strict";

import UIComponent from "../../../../base/UIComponent";
import classNames from "classnames";
import BannerItem from "../../banner/BannerItem";
import BannerManager from "../../banner/BannerManager";

const EVENTS = {
    ADD_BANNER: "DetailPanelBanners::addBanner",
    REMOVE_BANNER: "DetailPanelBanners::removeBanner"
};

const NAMESPACE = "mpageui-DetailPanelBanners";

/**
 * The DetailPanelBanners class.
 * @class DetailPanelBanners
 */
class DetailPanelBanners extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            addBannerEventName: EVENTS.ADD_BANNER,
            classNames: "",
            removeBannerEventName: EVENTS.REMOVE_BANNER
        };
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                bannerManager: new BannerManager({
                    addBannerEventName: this.getProp("addBannerEventName"),
                    removeBannerEventName: this.getProp("removeBannerEventName"),
                    itemDisplayMode: BannerItem.DISPLAYMODE.COMPACT
                })
            }
        ];
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mappedChildren) {
        return el(
            "div",
            {
                class: classNames({
                    [ NAMESPACE ]: true,
                    [ props.classNames ]: (props.classNames !== "")
                })
            },
            [
                mappedChildren.bannerManager.render()
            ]
        );
    }
}

DetailPanelBanners.EVENTS = EVENTS;

export default DetailPanelBanners;
