// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/progressive-disclosure/InlineDropdown.js
import UIComponent from "../../../base";
import classNames from "classnames";

const NAMESPACE = "mpageui-InlineDropDown";
const disclosureElementClass = "mpageui-disclosure-element";
const MAX_HEIGHT = 200;

class InlineDropDown extends UIComponent {
    initialProps() {
        return {
            classNames: "",
            content: [],
            maxHeight: MAX_HEIGHT,
            isVisible: false
        };
    }

    propChangeHandlers() {
        return {
            content: (content) => {
                this.replaceAllChildren(content);
            }
        };
    }

    view(el, props, children, mapped) {
        return props.isVisible ? el(
            "div",
            {
                class: classNames(NAMESPACE, {
                    [ props.classNames ]: !!props.classNames,
                    [ disclosureElementClass ]: true
                }),
                style: {
                    "max-height": `${props.maxHeight}px`
                }
            },
            this.renderChildren()
        ) : null;
    }
}

export default InlineDropDown;
