// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/search/bar/SearchTextClear.js
"use strict";
import UIComponent from "../../../../base/UIComponent";
import ClearIcon from "../../../atomic/icon/svg/category/content/Clear";
import classNames from "classnames";

const EVENTS = {
    CLEAR: "SearchTextClear::click"
};

const hideIconClass = "is-hidden";

/**
 * Renders the Clear icon that triggers an event to clear text in the search textBox
 */
class SearchTextClear extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            defaultClassNames: "mpageui-SearchTextClear-icon",
            isClearIconVisible: false,
            clickEventName: EVENTS.CLEAR
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mChildren) {
        return el(
            "a",
            {
                class: classNames({
                    [ props.defaultClassNames ]: true,
                    [ hideIconClass ]: !props.isClearIconVisible
                }),
                onclick: (event) => this.emit(this.getProp("clickEventName"), this, event)
            },
            [
                new ClearIcon().render()
            ]
        );
    }
}

SearchTextClear.EVENTS = EVENTS;
export default SearchTextClear;
