// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/search/manager/const.js
/**
 * Child name constants for SearchManager.
 * @type {{SEARCH_BAR: string, SEARCH_PANE: string, SEARCH_POPUP: string}}
 */
const CHILDREN = {
    SEARCH_BAR: "searchBar",
    SEARCH_PANE: "searchPane",
    SEARCH_POPUP: "searchPopup"
};

/**
 * Constant for default minimum length of search string for which control emits the input event.
 * @type {{DEFAULT_SEARCH_STRING_LENGTH: number}}
 */
const DEFAULT_SEARCH_STRING_LENGTH = 2;

/**
 * Public events for the SearchManager.
 * @type {{INPUT: string, SUBMIT: string, SELECTION_CHANGE: string}}
 */
const EVENTS = {
    FOCUS: "SearchManager::focus",
    BLUR: "SearchManager::blur",
    INPUT: "SearchManager::input",
    SUBMIT: "SearchManager::submit",
    SELECTION_CHANGE: "SearchManager::selectionChange"
};

const NAMESPACE = "mpageui-Search";

/**
 * Private events for the SearchManager.
 * @type {{DESKTOP: {FOCUS: string, BLUR: string}}}
 */
const PRIVATE_EVENTS = {
    DESKTOP: {
        FOCUS: "__SearchManager::Desktop::SearchBar::focus__",
        BLUR: "__SearchManager::Desktop::SearchBar::blur__"
    },
    MOBILE: {
        FOCUS: "__SearchManager::Mobile::SearchBar::focus__"
    }
};

export {
    CHILDREN,
    DEFAULT_SEARCH_STRING_LENGTH,
    EVENTS,
    NAMESPACE,
    PRIVATE_EVENTS
};
