// MPageFusion/testing/src/pages/controls/atomic/textbox/TextBox.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _seleniumWebdriver = require("selenium-webdriver");

var _UIComponent = require("../../../UIComponent.page");

var _UIComponent2 = _interopRequireDefault(_UIComponent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-TextBox";

/**
 * The TextBoxPage class
 * @class module:MPageFusion/testing.atomic.textbox.TextBoxPage
 *
 * @example
 * const textbox = new TextBoxPage(driver);
 * textbox.get("placeholder")
 *     .then((value) => {
 *         assert(value === "Enter a Value");
 *    });
 */

var TextBoxPage = function (_UIComponentPage) {
    _inherits(TextBoxPage, _UIComponentPage);

    function TextBoxPage() {
        _classCallCheck(this, TextBoxPage);

        return _possibleConstructorReturn(this, (TextBoxPage.__proto__ || Object.getPrototypeOf(TextBoxPage)).apply(this, arguments));
    }

    _createClass(TextBoxPage, [{
        key: "enterText",


        /**
         * sendkeys to the textbox
         * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
         * @param {String} text - the input text
         * @returns {Promise} resolves when the action is completed
         */
        value: function enterText(text) {
            return this.getRootElement().then(function (element) {
                return element.sendKeys(text);
            });
        }

        /**
         * the value in the textbox
         * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
         * @returns {Promise<String>} display text
         */

    }, {
        key: "clearText",


        /**
         * clears text from the textbox
         * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
         * @returns {Promise} resolves when the action is completed
         */
        value: function clearText() {
            return this.getRootElement().then(function (element) {
                return element.clear();
            });
        }

        /**
         * click the textbox
         * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
         * @returns {Promise} resolves when the action is completed
         */

    }, {
        key: "focus",
        value: function focus() {
            return this.getRootElement().then(function (element) {
                return element.click();
            });
        }

        /**
         * whether the textbox is focused
         * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
         * @returns {Promise<boolean>} isFocused
         */

    }, {
        key: "blur",


        /**
         * Whether the textbox is blur
         * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
         * @returns {Promise} resolves when the action is completed
         */
        value: function blur() {
            return this.getRootElement().then(function (element) {
                return element.sendKeys(_seleniumWebdriver.Key.TAB);
            });
        }
    }, {
        key: "elementSelector",
        get: function get() {
            return ROOT_SELECTOR;
        }

        /**
         * the placeholder
         * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
         * @returns {Promise<String>} display placeholder
         */

    }, {
        key: "placeholder",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("placeholder");
            });
        }

        /**
         * whether the textbox is disabled
         * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
         * @returns {Promise<boolean>} isDisabled
         */

    }, {
        key: "isDisabled",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("disabled");
            }).then(function (attribute) {
                return Boolean(attribute);
            });
        }

        /**
         * whether the textbox is read only
         * @memberOf module:MPageFusion/testing.atomic.textbox.TextBoxPage
         * @returns {Promise<boolean>} isReadOnly
         */

    }, {
        key: "isReadOnly",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("readonly");
            }).then(function (attribute) {
                return Boolean(attribute);
            });
        }
    }, {
        key: "value",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("value");
            });
        }
    }, {
        key: "isFocused",
        get: function get() {
            var _this2 = this;

            return this.getRootElement().then(function (element) {
                return _this2.driver.switchTo().activeElement().then(function (activeElement) {
                    return _seleniumWebdriver.WebElement.equals(element, activeElement);
                }).catch(function () {
                    return false;
                });
            });
        }
    }]);

    return TextBoxPage;
}(_UIComponent2.default);

exports.default = TextBoxPage;
