// MPageFusion/testing/src/pages/controls/atomic/grid/CardLayout.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _List = _interopRequireDefault(require("../../../helpers/List"));

var _CardCell = _interopRequireDefault(require("./CardCell.page"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-CellGrid";
var ROW_SELECTOR = ".mpageui-CellGrid-row";
/**
 * The CardLayoutPage class
 * @class module:MPageFusion/testing.atomic.grid.CardLayoutPage
 *
 * @example
 * const cardLayout = new CardLayoutPage(driver);
 */

var CardLayoutPage =
/*#__PURE__*/
function (_UIComponentPage) {
  _inheritsLoose(CardLayoutPage, _UIComponentPage);

  function CardLayoutPage() {
    return _UIComponentPage.apply(this, arguments) || this;
  }

  _createClass(CardLayoutPage, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * @memberOf module:MPageFusion/testing.atomic.grid.CardLayoutPage
     * @returns {List<module:MPageFusion/testing.atomic.grid.CardCellPage>} returns list of children
     */

  }, {
    key: "cellDetails",
    get: function get() {
      return (0, _List["default"])({
        list: this,
        itemPage: _CardCell["default"],
        selectorList: this.selectorList,
        containerSelector: ROW_SELECTOR
      });
    }
  }]);

  return CardLayoutPage;
}(_UIComponent["default"]);

exports["default"] = CardLayoutPage;
