// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/flowsheet/headers/composite/CellContainer.js
/**
 * @fileoverview TableHeader class
 * @author Gyandeep Singh
 */

"use strict";

import UIComponent from "../../../../../base";
import classNames from "classnames";

const primaryCellClass = "mpageui-Flowsheet-compositeHeader-primaryCell";
const secondaryCellClass = "mpageui-Flowsheet-compositeHeader-secondaryCell";

/**
 * Creates the flex style for the secondary cells of the header
 * @param {int} cnt - Column span number
 * @returns {string} flex css style string
 * @private
 */
const getFlexStyle = (cnt) => `flex: ${cnt} 1 0; -webkit-box-flex: ${cnt}; -ms-flex: ${cnt} 1 0; -ms-flex-positive: ${cnt};`;

/**
 * Constructor for CellContainer
 * @extends UIComponent
 */
export default class CellContainer extends UIComponent {
    /**
     * @inheritdoc
     */
    initialProps() {
        return {
            defaultClassNames: "mpageui-Flowsheet-compositeHeader-cell-wrapper",
            classNames: "",
            primary: {},
            secondary: []
        };
    }

    /**
     * @inheritdoc
     */
    view(el, props, children) {
        return el(
            "div",
            {
                class: classNames({
                    [ props.defaultClassNames ]: true,
                    [ props.classNames ]: props.classNames.length > 0
                }),
                style: getFlexStyle(props.secondary.length)
            },
            [
                el("div", {
                    class: primaryCellClass
                }, props.primary.display),
                el("div", {
                    class: secondaryCellClass
                }, props.secondary.map((cell) => el("div", {}, cell.display)))
            ]
        );
    }
}
