// eslint-plugin-mpages/index.js
'use strict';

module.exports = {
	configs: {
		recommended: {
			extends: "cerner",
			rules: {
				// Turn off import related rules, arrow-body-style, no-plusplus and no-underscore-dangle
				"import/no-extraneous-dependencies": "off",
				"import/no-unresolved": "off",
				"import/extensions": "off",
				"import/first": "off",
				"import/no-absolute-path": "off",
				"arrow-body-style": "off",
				"no-plusplus": "off",
				"no-underscore-dangle": "off",
				"linebreak-style": "off",

				// Following rules are updated to warning and custom options have been set
				"brace-style": ["error", "1tbs", { "allowSingleLine": true }],
				"max-len": ["error", 120, 2, { "ignoreComments": true }],
				"indent": ["error", 4, { "SwitchCase": 1 }],
				"quotes": ["error", "double"],

                // Following rules are error but with custom options
				"no-constant-condition": ["error", { "checkLoops": false }],
				"comma-dangle": ["error", "never"],
				"no-param-reassign": ["error", { "props": false }],
				"class-methods-use-this": ["error", {
					"exceptMethods": [
						"initialProps",
						"initialState",
						"propChangeHandlers",
						"createChildren",
						"beforeCreate",
						"afterCreate",
						"beforeRender",
						"view",
						"afterRender",
						"getSvg",
						"getRenderedBody",
						"elementSelector",
						"_name"                        
					]
				}],
				"no-unused-vars": ["error", {
					"vars": "all",
					"args": "none"
				}],
				"new-cap": [
					"error",
					{
						"capIsNewExceptions": [
							"DiscernObjectFactory",
							"SetPopupStringProp",
							"SetPopupDoubleProp",
							"SetPopupBoolProp",
							"LaunchPopup",
							"LaunchIView",
							"FilterRefresh",
							"InitiateGapCheck",
							"ToObject",
							"OpenDynDocByWorkflowId",
							"OpenNewDocumentByReferenceTemplateIdAndNoteType",
							"MP_Core.AppUserPreferenceManager.Initialize",
							"MP_Core.AppUserPreferenceManager.GetPreferences",
							"MP_Core.AppUserPreferenceManager.SetPreferences",
							"MP_Core.AppUserPreferenceManager.UpdateAllCompPreferences",
							"MP_Core.GetViewableEncntrs",
							"MP_Util.AddCookieProperty",
							"MP_Util.CreateTimer",
							"MP_Util.GetCookieProperty",
							"MP_Util.WriteCookie",
							"MP_Util.Doc.RunAccordian",
							"MP_Util.Doc.SetupCompFilters",
							"MP_Util.GetCompObjById",
							"MP_Util.GetComponentArray",
							"MP_Util.LogInfo",
							"MP_Util.LogScriptCallInfo",
							"MP_Util.ReleaseRequestReference",
							"MP_Util.CreateParamArray",
							"MP_Util.LoadCodeListJSON",
							"MP_Util.GetCodeByMeaning",
							"MP_Util.GetValueFromArray"
						]
					}
				]
			},
			parserOptions: {
				ecmaVersion: 2018,
				sourceType: "module",
				ecmaFeatures: {
					jsx: true
				}
			},
			env: {
				"browser": true,
				"jasmine": true,
				"node": true
			}
		}
	}
}
