// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/flowsheet/headers/composite/CompositeHeader.js
/**
 * @fileoverview TableHeader class
 * @author Gyandeep Singh
 */

"use strict";

import UIComponent from "../../../../../base";
import CellContainer from "./CellContainer";
import classNames from "classnames";
import { getFlowsheetMinWidth } from "./../../helpers";

/**
 * Constructor for CompositeHeader
 * @extends UIComponent
 */
export default class CompositeHeader extends UIComponent {
    /**
     * @inheritdoc
     */
    initialProps() {
        return {
            defaultClassNames: "mpageui-Flowsheet-compositeHeader",
            classNames: "",
            columns: [],
            minWidth: "auto"
        };
    }

    /**
     * @inheritdoc
     */
    createChildren() {
        return this.getProp("columns").map((column) => new CellContainer({
            classNames: column.classNames ? column.classNames : "",
            primary: column.primary,
            secondary: column.secondary
        }));
    }

    /**
     * Calculates the minimum width and returns it
     * @param {int} columnMinWidth - Minimum column width
     * @returns {string} minimum width in px
     */
    getMinWidth(columnMinWidth) {
        return getFlowsheetMinWidth(
            this.getProp("columns")
                .reduce((colCnt, { secondary }) => colCnt + secondary.length, 0), columnMinWidth);
    }

    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return {
            columns: () => {
                this.replaceAllChildren(this.createChildren());
            }
        };
    }

    /**
     * @inheritdoc
     */
    view(el, props, children) {
        return el(
            "div",
            {
                class: classNames({
                    [ props.defaultClassNames ]: true,
                    [ props.classNames ]: props.classNames.length > 0
                }),
                style: {
                    "min-width": props.minWidth
                }
            },
            [
                el("div", {
                    class: classNames({
                        "mpageui-Flowsheet-row-cell": true,
                        "mpageui-Flowsheet-row-cell--header": true
                    })
                }, "\u00a0"),
                ...this.renderChildren()
            ]
        );
    }
}
