// gaia-plugin-unit-testing/src/plugin.js
const commands = require("./commands");
const route = require("./route");
/**
 *  NOTES:
 *  Figure out how to exit CI mode without closing the server - not sure it's possible right now
 *  Running tests from cmd line - one setup exits after run and other doesn't with both set to SingleRun?
 *  Buggy issues with FormSelect: sometimes calls readExternals with placeholder and sometimes the targets are not loaded into select on initial server start
 **/

module.exports = {
    id: "unit-testing",
    commands: {
        run: {
            description: "Runs Jasmine unit tests within an artifact using Karma",
            handler: commands.run,
            args: {
                cli: "If set to true, the tests run in a headless browser and exit with status 1 on failure"
            }
        },
        readResults: {
            description: "Reads the JSON file output from the karma-plugin",
            handler: commands.readResults,
            args: {}
        },
        saveConfig: {
            description: "Writes out the config to the gaia.json file",
            handler: commands.saveConfig,
            args: {
                settings: "The object of config settings to be saved"
            }
        },
        readConfig: {
            description: "Reads the config from the gaia.json file",
            handler: commands.readConfig,
            args: {}
        },
        checkCodeCoverage: {
            description: "Checks the evidence folder for the code coverage report",
            handler: commands.checkCodeCoverage,
            args: {}
        }
    },
    routes: {
        coverage: {
            description: "Serves the code coverage when unit tests are ran",
            handler: route.coverage
        }
    },
    markdown: {
        ui: "docs/guide.md"
    },
    ui: "dist/ui.js"
};
