// gaia-plugin-contentserver/src/ui/gaia-plugin-contentserver/FolderList.js
import * as Fusion from "MPageFusion";
import AddFolder from "./AddFolder";
const Button = Fusion.atomic.button.Button;

const Table = Fusion.atomic.table.Table;

const DELETE_CLICK = "FolderList::DeleteClick";

export default class FolderList extends Fusion.UIComponent {

    initialProps() {
        return {
            folders: {},
            deleteEventName: "FolderList::Delete"
        };
    }

    propChangeHandlers() {
        return {
            folders: (folders) =>
                this.getChild("table").setProp(
                    "rows",
                    Object.keys(folders).map(
                        (alias) => ({
                            key: alias,
                            data: [
                                { display: alias },
                                { display: folders[alias] },
                                { content: [
                                    new Button({
                                        display: "Delete",
                                        clickEventName: DELETE_CLICK,
                                        _alias: alias
                                    })
                                ]}
                            ]
                        })
                    )
                )
        };
    }

    afterCreate() {
        this.on(DELETE_CLICK, (button) => {
            this.emit(this.getProp("deleteEventName"), button.getProp("_alias"));
        });
    }

    createChildren() {
        return [
            {
                table: new Table({
                    columns: [
                        { display: "Alias", key: "alias" },
                        { display: "Path", key: "path" },
                        { display: "", key: "actions"}
                    ]
                })
            }
        ];
    }

    view(el, props, children, mChildren) {
        return this.renderChildren();
    }
}
