// gaia-plugin-contentserver/src/ui/gaia-plugin-contentserver/index.js
import * as Fusion from "MPageFusion";
import FolderList from "./FolderList";
import AddFolder from "./AddFolder";
import { runCommand, routeUrl } from "Gaia";

const { forwardProp } = MPageFusion.helpers.props;
const BannerItem = Fusion.composite.banner.BannerItem;


const ADD_EVENT = "ContentServer::Add";
const DELETE_EVENT = "ContentServer::Delete";

const retrieveFolders = (control) =>
      runCommand(
          "contentserver",
          "allFolders"
      ).then(
          (folders) =>
              control
              .setProp("folders", folders)
              .update()
      );

class ContentServerUI extends Fusion.UIComponent {

    initialProps() {
        return {
            folders: {}
        };
    }

    propChangeHandlers() {
        return {
            folders: forwardProp(this, "folders", "folderList")
        };
    }

    createChildren() {
        return [
            {
                banner: new BannerItem({
                    title: "Content will be served from",
                    description: `${routeUrl("contentserver", "content")}/[alias]`,
                    enableDismissClick: false
                })
            },
            {
                addFolder: new AddFolder({
                    addEventName: ADD_EVENT
                })
            },
            {
                folderList: new FolderList({
                    deleteEventName: DELETE_EVENT
                })
            }
        ];
    }

    afterCreate() {
        retrieveFolders(this);

        this.on(
            ADD_EVENT,
            (folder) =>
                runCommand(
                    "contentserver",
                    "addFolder",
                    folder
                ).then(
                    () => retrieveFolders(this)
                )
        );

        this.on(
            DELETE_EVENT,
            (alias) =>
                runCommand(
                    "contentserver",
                    "removeFolder",
                    {
                        alias: alias
                    }
                ).then(
                    () => retrieveFolders(this)
                )
        );
    }

    view(el) {
        return el("div", { class: "gaia-root-section" }, this.renderChildren());
    }
};

export default {
	  label: "Content Server",
	  control: ContentServerUI,
    showHelp: true
};
