// MPageFusion/testing/src/pages/controls/atomic/content-panel/ContentPanel.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _seleniumWebdriver = require("selenium-webdriver");

var _UIComponent = require("../../../UIComponent.page");

var _UIComponent2 = _interopRequireDefault(_UIComponent);

var _ContentHeader = require("../header/content-header/ContentHeader.page");

var _ContentHeader2 = _interopRequireDefault(_ContentHeader);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-ContentPanel";
var BODY_SELECTOR = ".mpageui-ContentPanel-body";
var THEME = {
    OUTSIDE_RECORDS: "mpageui-u-bg-outside-records"
};

/**
 * The ContentPanelPage class
 * @class module:MPageFusion/testing.atomic.contentPanel.ContentPanelPage
 *
 * @example
 * const contentPanel = new ContentPanelPage(driver,"");
 * contentPanel.click()
 */

var ContentPanelPage = function (_UIComponentPage) {
    _inherits(ContentPanelPage, _UIComponentPage);

    function ContentPanelPage(driver, args) {
        _classCallCheck(this, ContentPanelPage);

        var _this = _possibleConstructorReturn(this, (ContentPanelPage.__proto__ || Object.getPrototypeOf(ContentPanelPage)).call(this, driver, args));

        _this._contentHeader = new _ContentHeader2.default(driver, {
            selectorList: _this.selectorList,
            useExactSelector: true
        });
        return _this;
    }

    _createClass(ContentPanelPage, [{
        key: "click",


        /**
         * toggle the content panel body on click of header
         * @memberOf module:MPageFusion/testing.atomic.contentPanel.ContentPanelPage
         * @returns {Promise<string>} click action
         */
        value: function click() {
            return this._contentHeader.leftAction.click();
        }

        /**
         * whether or not the content panel is collapsed
         * @memberOf module:MPageFusion/testing.atomic.contentPanel.ContentPanelPage
         * @returns {Promise<boolean>} isCollapsed
         */

    }, {
        key: "collapse",


        /**
         * collapses the content panel
         * @memberOf module:MPageFusion/testing.atomic.contentPanel.ContentPanelPage
         * @returns {Promise} - resolves when the action has completed
         */
        value: function collapse() {
            var _this2 = this;

            return this.isCollapsed.then(function (isCollapsed) {
                if (isCollapsed) {
                    throw new Error("ContentPanel not expanded");
                } else {
                    return _this2._contentHeader.leftAction.click();
                }
            });
        }

        /**
         * expands the content panel
         * @memberOf module:MPageFusion/testing.atomic.contentPanel.ContentPanelPage
         * @returns {Promise} - resolves when the action has completed
         */

    }, {
        key: "expand",
        value: function expand() {
            var _this3 = this;

            return this.isCollapsed.then(function (isCollapsed) {
                if (!isCollapsed) {
                    throw new Error("ContentPanel not collapsed");
                } else {
                    return _this3._contentHeader.leftAction.click();
                }
            });
        }

        /**
         * the contentPanel theme
         * @memberOf module:MPageFusion/testing.atomic.contentPanel.ContentPanel
         * @returns {Promise<string>} theme
         */

    }, {
        key: "elementSelector",
        get: function get() {
            return ROOT_SELECTOR;
        }

        /**
         * gets the content panel body selector list
         * @memberOf module:MPageFusion/testing.atomic.contentPanel.ContentPanelPage
         * @returns {Array} selector list
         */

    }, {
        key: "bodySelectorList",
        get: function get() {
            return this.selectorList.concat(BODY_SELECTOR);
        }

        /**
         * gets the content panel footer selector list
         * @memberOf module:MPageFusion/testing.atomic.contentPanel.ContentPanelPage
         * @returns {Array} selector list
         */

    }, {
        key: "footerSelectorList",
        get: function get() {
            return this.selectorList;
        }

        /**
         * gets the content panel header display
         * @memberOf module:MPageFusion/testing.atomic.contentPanel.ContentPanelPage
         * @returns {Promise<string>} header display
         */

    }, {
        key: "headerDisplay",
        get: function get() {
            return this._contentHeader.display;
        }
    }, {
        key: "isCollapsed",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.findElement(_seleniumWebdriver.By.css(BODY_SELECTOR)).then(function () {
                    return false;
                }).catch(function () {
                    return true;
                });
            });
        }
    }, {
        key: "theme",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("class");
            }).then(function (classNames) {
                return Object.values(THEME).find(function (backgroundTheme) {
                    return classNames.includes(THEME.OUTSIDE_RECORDS);
                }) || "";
            });
        }
    }]);

    return ContentPanelPage;
}(_UIComponent2.default);

exports.default = ContentPanelPage;
