// MPageFusion/build/webpack/helpers/babel.js
"use strict";

const baseConfig = (type, version = "DEV") => ({
    loader: "babel-loader",
    options: {
        presets: [
            ["babel-preset-mpage", { 
                modules: false
            }]
        ],
        plugins: [
            ["@babel/plugin-transform-runtime", {
                "helpers": true,
                "corejs": type === "DIST" ? 2 : false,
                "regenerator": type === "DIST"
            }],
            ["babel-plugin-minify-replace", {
                "replacements": [
                    {
                        "identifierName": "VERSION",
                        "replacement": {
                            type: "StringLiteral",
                            value: version
                        }
                    },
                    {
                        "identifierName": "__DEV__",
                        "replacement": {
                            type: "booleanLiteral",
                            value: type === "DEV"
                        }
                    }
                ]
            }]
        ]
    }
});

module.exports = {
    baseConfig
};
