// fusion-test-helpers/build/karma.config.js
const path = require('path');

module.exports = function(config) {
    config.set({
        basePath: '..',
        browsers: ["Chrome"],
        frameworks: ["jasmine"],
        autoWatch: true,
        files: [
            "./build/tests.webpack.js"
        ],
        preprocessors: {
            "./build/tests.webpack.js": ['webpack']
        },
        webpack: {
            module: {
                rules: [
                    {
                        test: /\.js$/,
                        include: path.resolve("./src/testHelpers.js"),
                        exclude: /(node_modules)/,
                        use: {
                            loader: "babel-loader",
                            options: {
                                presets: [["es2015"]],
                                plugins: [
                                    [
                                        "istanbul",
                                        {
                                            "exclude": [
                                                "**/__test__/**/*.js",
                                            ]
                                        }
                                    ]
                                ]
                            }
                        }
                    }
                ]
            }
        },
        reporters: [ 'progress', 'coverage-istanbul' ],
        coverageIstanbulReporter: {
            dir: ".coverage",
            reports: [ 'html' ],
            fixWebpackSourcePaths: true
        }
    });
};
