// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/footer/FooterBase.js
/**
 * @fileoverview FooterBase class
 * @author Gyandeep Singh
 */

"use strict";

import UIComponent from "../../../base";
import classNames from "classnames";

/**
 * Default css classes
 * @type {{default: string, lSection: string, rSection: string, canvasStyle: string, anchorStyle: string}}
 */
const defaultCssClasses = {
    default: "mpageui-FooterBase",
    lSection: "mpageui-FooterBase-lSection",
    rSection: "mpageui-FooterBase-rSection",
    bisected: "mpageui-FooterBase--bisected",
    notBisected: "mpageui-FooterBase--notBisected",
    canvasStyle: "mpageui-FooterBase-canvas",
    anchorStyle: "mpageui-FooterBase-anchor",
    noLeftPad: "mpageui-FooterBase--no-left-pad"
};

/**
 * Render the sections of the content bar
 * @param {Function} el - VDom creator
 * @param {UIComponent|undefined} child - UIComponent if present
 * @param {string} cssClass - CSS class
 * @returns {VDom} VDom for the section
 * @private
 */
const renderSection = (el, child, cssClass) => el(
    "div",
    {
        class: cssClass
    },
    child ? child.render() : ""
);

/**
 * FooterBase class.
 * @class FooterBase
 * @constructor
 */
export default class FooterBase extends UIComponent {
    /**
     * @inheritdoc
     */
    initialProps() {
        return {
            defaultClassNames: defaultCssClasses.default,
            classNames: "",
            lSection: null,
            lClassNames: "",
            rSection: null,
            rClassNames: "",
            isBisected: false,
            bgStyle: FooterBase.BG_STYLE.ANCHOR
        };
    }

    /**
     * @inheritdoc
     */
    propChangeHandlers() {
        return {
            lSection: (newComp) => newComp ? this.replaceMappedChild("lSection", newComp) : this.removeChild("lSection"),
            rSection: (newComp) => newComp ? this.replaceMappedChild("rSection", newComp) : this.removeChild("rSection")
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mChildren) {
        return el(
            "div",
            {
                class: classNames({
                    [ props.defaultClassNames ]: true,
                    [ props.classNames ]: props.classNames.length > 0,
                    [ props.bgStyle ]: true,
                    [ defaultCssClasses.bisected ]: props.isBisected,
                    [ defaultCssClasses.notBisected ]: !props.isBisected
                })
            },
            [
                // left section
                renderSection(
                    el,
                    mChildren.lSection,
                    classNames({
                        [ defaultCssClasses.lSection ]: !!props.lSection,
                        [ props.lClassNames ]: props.lClassNames.length > 0
                    })
                ),
                // right section
                renderSection(
                    el,
                    mChildren.rSection,
                    classNames({
                        [ defaultCssClasses.rSection ]: !!props.rSection,
                        [ props.rClassNames ]: props.rClassNames.length > 0,
                        [ defaultCssClasses.noLeftPad ]: true
                    })
                )
            ]
        );
    }
}

/**
 * Default background styles
 * @type {{CANVAS: string, ANCHOR: string}}
 */
FooterBase.BG_STYLE = {
    CANVAS: defaultCssClasses.canvasStyle,
    ANCHOR: defaultCssClasses.anchorStyle
};
