// MPageFusion/testing/src/pages/controls/composite/combobox/Combobox.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _TextBox = _interopRequireDefault(require("../../atomic/textbox/TextBox.page"));

var _Button = _interopRequireDefault(require("../../atomic/button/Button.page"));

var _SearchListItem = _interopRequireDefault(require("../search/SearchListItem.page"));

var _List = _interopRequireDefault(require("../../../helpers/List"));

var _seleniumWebdriver = require("selenium-webdriver");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-ComboBox";
var DROP_DOWN_SELECTOR = ".mpageui-InlineDropDown";
var SEARCH_LIST_SELECTOR = ".mpageui-SearchList";
/**
 * The comboboxPage class
 * @class module:MPageFusion/test.composite.combobox.Comboboxpage
 *
 * @private
 */

var ComboboxPage =
/*#__PURE__*/
function (_UIComponentPage) {
  _inheritsLoose(ComboboxPage, _UIComponentPage);

  function ComboboxPage(driver, options) {
    var _this;

    _this = _UIComponentPage.call(this, driver, options) || this;
    _this._button = new _Button["default"](driver, {
      selectorList: _this.selectorList
    });
    _this._textBox = new _TextBox["default"](driver, {
      selectorList: _this.selectorList
    });
    return _this;
  }

  var _proto = ComboboxPage.prototype;

  /**
   * focuses the textbox
   * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
   * @returns {Promise} resolves when the action is completed
   */
  _proto.focus = function focus() {
    return this._textBox.focus();
  }
  /**
   * Whether the combobox is blur
   * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
   * @returns {Promise} resolves when the action is completed
   */
  ;

  _proto.blur = function blur() {
    return this._textBox.blur();
  }
  /**
   * enters text to the combobox
   * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
   * @param {String} text -the input text
   * @returns {Promise} resolves when the action is completed
   */
  ;

  _proto.enterText = function enterText(text) {
    return this._textBox.enterText(text);
  }
  /**
   * clears text from the combobox textbox
   * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
   * @returns {Promise} resolves when the action is completed
   */
  ;

  _proto.clearText = function clearText() {
    return this._textBox.clearText();
  }
  /**
   * clicks the dropdown icon button in the combobox bar
   * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
   * @returns {Promise} resolves when the action is completed
   */
  ;

  _proto.click = function click() {
    return this._button.click();
  };

  _createClass(ComboboxPage, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * whether the combobox bar is disabled
     * @memberOf module:MPageFusion/testing.composite.comboxbox.FormComboboxPage
     * @returns {Promise<boolean>} true or false
     */

  }, {
    key: "isDisabled",
    get: function get() {
      return Promise.all([this._textBox.isDisabled, this._button.isDisabled]).then(function (results) {
        return Boolean(results[0] && results[1]);
      });
    }
    /**
     * whether the search list is displayed
     * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
     * @returns {Promise<boolean>} true or false
     */

  }, {
    key: "isOpen",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.findElement(_seleniumWebdriver.By.css(DROP_DOWN_SELECTOR));
      }).then(function () {
        return true;
      })["catch"](function () {
        return false;
      });
    }
    /**
     * returns placeholder text from the combobox
     * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
     * @returns {Promise<string>} placeholder text
     */

  }, {
    key: "placeholder",
    get: function get() {
      return this._textBox.placeholder;
    }
    /**
     * the value in the textbox
     * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
     * @returns {Promise<String>} display text
     */

  }, {
    key: "value",
    get: function get() {
      return this._textBox.value;
    }
    /**
     * returns the item from the search list based on option(index)
     * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
     * @returns {List<module:MPageFusion/testing.composite.search.SearchPage.searchList>} searchList
     */

  }, {
    key: "option",
    get: function get() {
      return (0, _List["default"])({
        list: this,
        itemPage: _SearchListItem["default"],
        selectorList: this.selectorList,
        containerSelector: DROP_DOWN_SELECTOR + " " + SEARCH_LIST_SELECTOR
      });
    }
    /**
     * whether the combo box is focused
     * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
     * @returns {Promise<boolean>} true or false
     */

  }, {
    key: "isFocused",
    get: function get() {
      return this._textBox.isFocused;
    }
  }]);

  return ComboboxPage;
}(_UIComponent["default"]);

exports["default"] = ComboboxPage;
