// gaia-plugin-whats-up-doc/src/ui/js/control/Guide.js
/**
 * @module whatsupdoc/ui/control/Guide
 * @requires MPageFusion
 * @requires Gaia
 * @requires module:whatsupdoc/ui/control/Marked
 */
import { composite, UIComponent } from "MPageFusion";
import { runCommand } from "Gaia";
import Marked from "./Marked";

const {
    navigation: {
        TabControl
    }
} = composite;

/**
 * Strip the extension off the file name.
 *
 * @private
 * @function module:whatsupdoc/ui/control/Guide~stripExtension
 * @param {String} file - full name of the file
 * @return {String} Returns file name without extension
 */
const stripExtension = file => file.slice(0, -3);

/**
 * @alias module:whatsupdoc/ui/control/Guide
 */
class Guide extends UIComponent {
    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                tabs: new TabControl({
                    orientation: TabControl.ORIENTATION.VERTICAL,
                    tabContentClassNames: "mpageui-u-margin-left-relative-loose"
                })
            }
        ];
    }

    /**
     * @inheritDoc
     */
    afterCreate() {
        // Get the markdown guide files and convert to tabs
        runCommand("whats-up-doc", "getGuide", {})
            .then(files => this
                .getChild("tabs")
                .setProp("content",
                    files.map((file, index) => ({
                        id: `${index}`,
                        display: stripExtension(file.name),
                        view: new Marked({
                            value: file.text
                        })
                    }))
                )
            )
            .catch(console.error);
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mappedChildren) {
        return mappedChildren.tabs.getProp("content").length
            ? this.renderChildren()
            : null;
    }
}

export default Guide;
