// gaia-plugin-whats-up-doc/src/ui/js/control/Marked.js
/**
 * @module whatsupdoc/ui/control/Marked
 */

import { UIComponent } from "MPageFusion";
import classnames from "classnames";
import { runCommand } from "Gaia";
import marked from "marked";

const NAMESPACE = "wud--Marked";

/**
 * @name module:whatsupdoc/ui/control/Marked~updateInnerHTML~innerHTML
 * @function
 * @param {VDOM} vdom - A vdom object
 * @param {HTMLNode} vdom.dom - the DOM object of the vdom
 * @returns {undefined}
 */

/**
 * Curried function that takes in an html string and a VDOM object
 * and sets the innerHTML on the VDOM object.
 *
 * @inner
 * @private
 * @function
 * @param {string} html - An HTML string
 * @returns {innerHTML}
 */
const updateInnerHTML = html => ({ dom }) => dom.innerHTML = html;

/**
 * @alias module:whatsupdoc/ui/control/Marked
 */
class Marked extends UIComponent {
    /**
     * @inheritdoc
     */
    constructor(...args) {
        super(...args);
        this._marked = marked;
    }

    /**
     * @inheritdoc
     */
    initialProps() {
        return {
            markedOptions: {
                gfm: true,
                tables: true,
                breaks: false,
                pedantic: false,
                sanitize: false,
                smartLists: true,
                smartypants: false,
                langPrefix: 'lang-'
            },
            value: ""
        };
    }

    /**
     * @inheritdoc
     */
    initialState() {
        return {
            html: ""
        };
    }

    /**
     * @inheritdoc
     */
    dependentPropChangeHandlers() {
        return [
            ["markedOptions", "value", (markedOptions, value) => {
                // Set the options
                this._marked.setOptions(markedOptions);

                // Re-render the HTML with either the new options or the new value
                this.setState({
                    html: this._marked(value)
                });
            }]
        ];
    }

    /**
     * @inheritdoc
     */
    view(el, props, children, mappedChildren, { state }) {
        return el("section", {
            class: classnames(NAMESPACE),
            oncreate: updateInnerHTML(state.html),
            onupdate: updateInnerHTML(state.html)
        });
    }
}

export default Marked;
