// MPageFusion/build/karma/karma.config.js
const path = require("path");
const { withLocalI18n } = require("../webpack/helpers/i18n");

module.exports = function(config) {
    config.set({
        browsers: [ "PhantomJS" ],
        coverageReporter: {
            dir: "../../.coverage",
            reporters: [
                { type: "html", subdir: "html" },
                { type: "cobertura", subdir: "." }
            ]
        },
        singleRun: true,
        browserNoActivityTimeout: 100000, // default: 10000; doing this for jenkins
        files: [
            "../../node_modules/babel-polyfill/dist/polyfill.js",
            "../webpack/tests.webpack.js"
        ],
        frameworks: [
            "jasmine"
        ],
        preprocessors: {
            "../webpack/tests.webpack.js": [ "webpack", "sourcemap" ]
        },
        reporters: [ "progress", "coverage" ],
        webpackMiddleware: {
            noInfo: true
        },
        webpack: withLocalI18n({
            cache: true,
            devtool: "inline-source-map",
            resolve: {
                alias: {
                    "cerner": path.resolve(__dirname, "..", "..", "src", "main", "js", "cerner")
                },
                modules: [
                    "node_modules"
                ]
            },
            module: {
                rules: [
                    {
                        test: /\.js$/,
                        include: path.resolve(__dirname, "../../src"),
                        exclude: /(node_modules)/,
                        use: [
                            {
                                loader: "babel-loader",
                                options: {
                                    presets: [["env"]],
                                    plugins: [
                                        ["transform-es2015-classes", {loose: true}],
                                        ["babel-plugin-minify-replace", {
                                            "replacements": [
                                                {
                                                    "identifierName": "VERSION",
                                                    "replacement": {
                                                        type: "StringLiteral",
                                                        value: "TEST",
                                                    }
                                                },
                                                {
                                                    "identifierName": "__DEV__",
                                                    "replacement": {
                                                        type: "booleanLiteral",
                                                        value: true
                                                    }
                                                }
                                            ]
                                        }],
                                        [ "istanbul", {
                                            "exclude": [
                                                "**/__tests__/**/*.js",
                                                "**/atomic/icon/svg/svgDefinitions.js",
                                                "**/base/EventEmitter.js",
                                                "**/base/devChecker.js"
                                            ]
                                        }]
                                    ]
                                }
                            }
                        ]
                    }
                ]
            }
        })
    });
};
