// gaia-plugin-whats-up-doc/src/ui/js/Body.js
/**
 * @module whatsupdoc/ui/Body
 *
 * @requires MPageFusion
 * @requires module:whatsupdoc/i18n
 * @requires module:whatsupdoc/ui/control/DocViewer
 * @requires module:whatsupdoc/ui/control/Guide
 */

import { composite, UIComponent } from "MPageFusion";
import i18n from "../../i18n/en_us";
import DocView from "./control/DocViewer";
import Guide from "./control/Guide";
import Config from "./control/Config";

const {
    navigation: {
        TabControl
    }
} = composite;

/**
 * @alias module:whatsupdoc/ui/Body
 */
class Body extends UIComponent {
    /**
     * @inheritdoc
     */
    createChildren() {
        return [
            {
                tabs: new TabControl({
                    content: [{
                        id: "docs",
                        display: i18n.tabs.docs,
                        view: new DocView()
                    }, {
                        id: "guide",
                        display: i18n.tabs.guide,
                        view: new Guide()
                    }, {
                        id: "config",
                        display: i18n.tabs.config,
                        view: new Config()
                    }]
                })
            }
        ];
    }

    /**
     * @inheritdoc
     */
    view(el, props, children, mappedChildren) {
        return mappedChildren.tabs.render();
    }
}

export default Body;
