// MPageFusion/build/karma/karma.config.nocoverage.js
const path = require("path");
const { withLocalI18n } = require("../webpack/helpers/i18n");

module.exports = function(config) {
    config.set({
        browsers: [ "PhantomJS" ],
        singleRun: true,
        files: [
            "../../node_modules/babel-polyfill/dist/polyfill.js",
            "../webpack/tests.webpack.js"
        ],
        frameworks: [
            "jasmine"
        ],
        preprocessors: {
            "../webpack/tests.webpack.js": [ "webpack" ]
        },
        reporters: [ "progress" ],
        webpackMiddleware: {
            noInfo: true
        },
        webpack: withLocalI18n({
            cache: true,
            devtool: "inline-source-map",
            resolve: {
                alias: {
                    "cerner": path.resolve(__dirname, "..", "..", "src", "main", "js", "cerner")
                },
                modules: [
                    "node_modules"
                ]
            },
            module: {
                rules: [
                    {
                        test: /\.js$/,
                        include: path.resolve(__dirname, "../../src"),
                        exclude: /(node_modules)/,
                        use: [
                            {
                                loader: "babel-loader",
                                options: {
                                    presets: [
                                        ["babel-preset-mpage", { modules: "commonjs" }]
                                    ],
                                    plugins: [
                                        ["babel-plugin-minify-replace", {
                                            "replacements": [
                                                {
                                                    "identifierName": "VERSION",
                                                    "replacement": {
                                                        type: "StringLiteral",
                                                        value: "TEST"
                                                    }
                                                },
                                                {
                                                    "identifierName": "__DEV__",
                                                    "replacement": {
                                                        type: "booleanLiteral",
                                                        value: true
                                                    }
                                                }
                                            ]
                                        }]
                                    ]
                                }
                            }
                        ]
                    }
                ]
            }
        })
    });
};
