// gaia-plugin-livedev/src/ui/gaia-plugin-livedev/launch-pad/LaunchButton.js
import { atomic, UIComponent } from "MPageFusion";
import { DISPATCH, REQUEST_LAUNCH } from "../actions";

/**
 * Creates the program launch button.
 * @class LaunchButton
 */
class LaunchButton extends UIComponent {
    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return {
            disabled: (disabled) => this.getChild("button").setProp("disabled", disabled)
        };
    }

    /**
     * @inheritDoc
     */
    afterCreate() {
        this.on(atomic.button.Button.EVENTS.CLICK, () => {
            this.emit(DISPATCH, {
                type: REQUEST_LAUNCH
            });
        })
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                button: new atomic.button.Button({
                    classNames: "gaia-plugin-LiveDev-LaunchButton",
                    display: "Launch"
                })
            }
        ]
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mapped) {
        return this.renderChildren();
    }
}

export default LaunchButton;
