// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/collapse-group/const.js
"use strict";

import Button from "../../atomic/button/Button";

/**
 * Enumeration for events which can be emitted by the CollapseGroup.
 * @type {{DROPDOWN_CLICK: string, MENU_SELECTION_CHANGE: string, MENU_TOGGLE: string, MENU_ACTION: string}}
 */
export const EVENTS = {
    DROPDOWN_CLICK: "CollapseGroup::Dropdown::click",
    MENU_SELECTION_CHANGE: "CollapseGroup::Menu::selectionChange",
    MENU_TOGGLE: "CollapseGroup::Menu::toggle",
    MENU_ACTION: "CollapseGroup::Menu::action"
};

/**
 * Enumeration for additional modifiers on CollapseGroup items.
 * @type {{ACTION: {STYLE: {SECONDARY: string}}, TOGGLE: {STYLE: {SECONDARY: string}}, SELECT_GROUP: {STYLE: {SECONDARY: string}}}}
 */
export const MODIFIERS = {
    ACTION: {
        STYLE: {
            SECONDARY: Button.TYPE.SECONDARY,
            PRIMARY: Button.TYPE.PRIMARY,
            LINK: Button.TYPE.LINK
        },
        SIZE: {
            THIN: Button.SIZE.THIN
        }
    },
    TOGGLE: {
        STYLE: {
            SECONDARY: Button.TYPE.SECONDARY
        },
        SIZE: {
            THIN: Button.SIZE.THIN
        }
    },
    SELECT_GROUP: {
        STYLE: {
            SECONDARY: Button.TYPE.SECONDARY
        }
    }
};

export const NAMESPACE = "mpageui-CollapseGroup";

export const TYPES = {
    ACTION: "action",
    DIVIDER: "divider",
    TOGGLE: "toggle",
    SELECT_GROUP: "selectGroup",
    MENU: "menu"
};

export const BUTTON_SIZE = {
    THIN: Button.SIZE.THIN
};
