// MPageFusion/testing/src/pages/controls/atomic/card/helper/cardHelper.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isIncludedClass = exports.displayHelper = exports.makeListHelper = exports.getElement = void 0;

var getElement = function getElement(root, contentSelector) {
  return root.getNestedElement(contentSelector);
};

exports.getElement = getElement;

var makeListHelper = function makeListHelper(list, itemPage, selectorList, containerSelector) {
  return {
    list: list,
    itemPage: itemPage,
    selectorList: selectorList,
    containerSelector: containerSelector
  };
};

exports.makeListHelper = makeListHelper;

var displayHelper = function displayHelper(root) {
  return root.getRootElement().then(function (element) {
    return element.getText();
  });
};

exports.displayHelper = displayHelper;

var isIncludedClass = function isIncludedClass(root, item) {
  return root.getRootElement().then(function (element) {
    return element.getAttribute("class");
  }).then(function (classNames) {
    return !!classNames.includes(item);
  });
};

exports.isIncludedClass = isIncludedClass;
