// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/select-group/CheckboxGroup.js
"use strict";

import SelectGroup from "./SelectGroup";
import Checkbox from "../../atomic/checkbox";
import { unique } from "../../../helpers/core/generators";

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------
const CheckboxGroupEvents = {
    SELECTION_CHANGE: "CheckboxGroup::selectionChange"
};

const namespace = "mpageui-CheckboxGroup";

/**
 * The CheckboxGroup class.
 * @class CheckboxGroup
 * @augments  SelectGroup
 */
export default class CheckboxGroup extends SelectGroup {
    /**
     * @inheritDoc
     */
    initialProps() {
        const uniqueFormName = unique(namespace);

        return Object.assign({}, super.initialProps(), {
            canDeselect: true,
            defaultClassNames: namespace,
            formName: uniqueFormName,
            isInline: false,
            isDisabled: false,
            multiSelect: true,
            selectionChangeEventName: CheckboxGroupEvents.SELECTION_CHANGE,
            itemGenerator: (props) => new Checkbox(Object.assign({},
                props, this._generateMappedPropResolver()), []) // eslint-disable-line no-underscore-dangle
        });
    }

    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return Object.assign({}, super.propChangeHandlers(), {
            formName: () => { this._runRefreshItems = true; },
            isDisabled: () => { this._runRefreshItems = true; },
            isInline: () => { this._runRefreshItems = true; }
        });
    }

    /**
     * Allows props of the CheckboxGroup to be mapped to each Checkbox
     * @returns {object} the mapped properties
     * @private
     */
    _generateMappedPropResolver() {
        const isDisabled = this.getProp("isDisabled");
        return Object.assign(
            {
                name: this.getProp("formName"),
                isInline: this.getProp("isInline")
            },
            // checkboxes can be individually disabled if isDisabled is set to null
            isDisabled !== null ? { isDisabled } : null
        );
    }
}

CheckboxGroup.EVENTS = CheckboxGroupEvents;
