// gaia-plugin-linter/src/ui/ReportBanner.js
const classNames = require('classnames');
import { UIComponent } from "MPageFusion";

/**
 * @class ReportBanner
 */
export default class ReportBanner extends UIComponent {

    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            title: "ESLint Report",
            defaultBannerClass: "linter-report-banner",
            bannerClasses: "linter-error-colors",
            summaryText: "99 problems",
            generatedDate: "when?",
            subtitleClass: "linter-report-subtitle"
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props) {
        return el(
            "div",
            { class: classNames({
                [props.defaultBannerClass]: true,
                [props.bannerClasses]: props.bannerClasses.length > 0
            }) },
            [
                el("h1", {}, props.title),
                el("p", { class: props.subtitleClass }, `${props.summaryText} - Generated on ${props.generatedDate}`)
            ]
        );
    }
}
