// gaia-plugin-injector/src/ui/gaia-plugin-injector/Settings.js
import * as Fusion from "MPageFusion";

const { forwardProp } = Fusion.helpers.props;
const FormTextBox = Fusion.composite.form.FormTextBox;

const TARGET_URL_CHANGE = "Internal::Settings::TargetUrlChange";
const PREFIX_CHANGE = "Internal::Settings::PrefixChange";

export default class Settings extends Fusion.UIComponent {
    initialProps() {
        return {
            targetUrl: "",
            prefix: "",
            targetUrlChangedEventName: "Settings::TargetUrlChange",
            prefixChangedEventName: "Settings::PrefixChange"
        };
    }

    propChangeHandlers() {
        return {
            targetUrl: forwardProp(this, "value", "targetUrl"),
            prefix: forwardProp(this, "value", "prefix")
        };
    }

    createChildren() {
        return [
            {
                targetUrl: new FormTextBox({
                    display: "Target URL",
                    classNames: "mpageui-u-flexgrid-col",
                    valueChangeEventName: TARGET_URL_CHANGE
                })
            },
            {
                prefix: new FormTextBox({
                    display: "Prefix",
                    classNames: "mpageui-u-flexgrid-col",
                    valueChangeEventName: PREFIX_CHANGE
                })
            }
        ];
    }

    afterCreate() {
        this.convertEventUsingProp(
            TARGET_URL_CHANGE,
            "targetUrlChangedEventName",
            (source, value) => [ this, value.value ]
        );

        this.convertEventUsingProp(
            PREFIX_CHANGE,
            "prefixChangedEventName",
            (source, value) => [ this, value.value ]
        );
    }

    view(el, props, children, mChildren) {
        return this.renderChildren();
    }
}
