// gaia-plugin-livedev/src/ui/gaia-plugin-livedev/driver-config/CCLProgramInput.js
import { UIComponent, composite } from "MPageFusion";
import { DISPATCH, CCL_PROGRAM_CHANGE } from "../actions";
const { form: { FormTextBox }} = composite;

/**
 * Events private to this control.
 * @type {{VALUE_CHANGE: string}}
 */
const PRIVATE_EVENTS = {
    VALUE_CHANGE: "__CCLProgramInput::valueChange__"
};

/**
 * Creates an input for the ccl program.
 * @class CCLProgramInput
 */
class CCLProgramInput extends UIComponent {
    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return {
            value: (value) => this.getChild("input").setProp("value", value)
        }
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                input: new FormTextBox({
                    helperHandler: () => "The name of the CCL program responsible for launching the application",
                    showHelperText: true,
                    display: "CCL Program",
                    valueChangeEventName: PRIVATE_EVENTS.VALUE_CHANGE
                })
            }
        ]
    }

    /**
     * @inheritDoc
     */
    afterCreate() {
        this.on(PRIVATE_EVENTS.VALUE_CHANGE, (src, payload) => {
            this.stopPropagation(PRIVATE_EVENTS.VALUE_CHANGE);
            this.emit(DISPATCH, {
                type: CCL_PROGRAM_CHANGE,
                payload: payload
            });
        });
    }

    /**
     * @inheritDoc
     */
    view(el, props, children) {
        return this.renderChildren();
    }
}

export default CCLProgramInput;
