// gaia-plugin-livedev/src/ui/gaia-plugin-livedev/driver-config/WebserviceURL.js
import { UIComponent, composite, helpers } from "MPageFusion";
import { DISPATCH, WEBSERVICE_CHANGE } from "../actions";

const FormTextBox = composite.form.FormTextBox;
const forwardProp = helpers.props.forwardProp;

const VALUE_CHANGE = "WebserviceURL::ValueChange";

export default class WebserviceURL extends UIComponent {
    propChangeHandlers() {
        return {
            value: forwardProp(this, "value", "input")
        };
    }

    createChildren() {
        return [
            {
                input: new FormTextBox({
                    helperHandler: () =>
                        "URL of the MPages Web Service if you are using one. Otherwise, leave it blank.",
                    showHelperText: true,
                    display: "Web Service URL",
                    valueChangeEventName: VALUE_CHANGE
                })
            }
        ];
    }

    afterCreate() {
        this.convertEvent(
            VALUE_CHANGE,
            DISPATCH,
            (src, payload) => [{
                type: WEBSERVICE_CHANGE,
                payload: payload.value
            }]
        );
    }

    view(el, props, children) {
        return this.renderChildren();
    }
}
