// gaia-plugin-config/src/plugin.js
const commands = require("./commands");

module.exports = {
    id: "config",
    commands: {
        read: {
            description: [
                "Reads a plugin configuration from disk. A single file can contain",
                "multiple plugin configurations. If the file does not exists, an empty",
                "object will be returned."
            ].join(" "),
            handler: commands.read,
            args: {
                plugin: {
                    description: "The ID of the plugin"
                },
                gaiaJson: {
                    description: [
                        "The full path of the file to read from. Defaults to the current",
                        "directory's gaia.json if not provided. "
                    ].join(" ")
                }
            }
        },
        readPrivate: {
            description: [
                "Reads a plugin configuration from disk. A single file can contain",
                "multiple plugin configurations. If the file does not exists, an empty",
                "object will be returned."
            ].join(" "),
            handler: commands.readPrivate,
            args: {
                plugin: {
                    description: "The ID of the plugin"
                },
                gaiaJson: {
                    description: [
                        "The full path of the file to read from. Defaults to the current",
                        "directory's gaiaPrivate.json if not provided. "
                    ].join(" ")
                }
            }
        },
        write: {
            description: [
                "Writes a plugin configuration into a file. A single file can contain",
                "multiple plugin configurations. If the file does not exists, a new one",
                "will be created."
            ].join(" "),
            handler: commands.write,
            args: {
                plugin: {
                    description: "The ID of the plugin"
                },
                config: {
                    description: "The configuration data to be written, usually in JSON format."
                },
                gaiaJson: {
                    description: [
                        "The full path of the file to read from. Defaults to the current",
                        "directory's gaia.json if not provided. "
                    ].join(" ")
                }
            }
        },
        writePrivate: {
            description: [
                "Writes a plugin configuration into a file. A single file can contain",
                "multiple plugin configurations. If the file does not exists, a new one",
                "will be created."
            ].join(" "),
            handler: commands.writePrivate,
            args: {
                plugin: {
                    description: "The ID of the plugin"
                },
                config: {
                    description: "The configuration data to be written, usually in JSON format."
                },
                gaiaJson: {
                    description: [
                        "The full path of the file to read from. Defaults to the current",
                        "directory's gaiaPrivate.json if not provided. "
                    ].join(" ")
                }
            }
        }
    },
    ui: ""
};
