// MPageFusion/testing/src/pages/controls/atomic/flowsheet/FlowsheetRow.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _UIComponent = require("../../../UIComponent.page");

var _UIComponent2 = _interopRequireDefault(_UIComponent);

var _FlowsheetCell = require("./FlowsheetCell.page");

var _FlowsheetCell2 = _interopRequireDefault(_FlowsheetCell);

var _List = require("../../../helpers/List");

var _List2 = _interopRequireDefault(_List);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-Flowsheet-row-wrapper";
var CELL_CONTAINER = ".mpageui-Flowsheet-row";

/**
 * The FlowsheetRowPage class
 * @alias module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
 * @class FlowsheetRowPage
 * @private
 */

var FlowsheetRow = function (_UIComponentPage) {
    _inherits(FlowsheetRow, _UIComponentPage);

    function FlowsheetRow() {
        _classCallCheck(this, FlowsheetRow);

        return _possibleConstructorReturn(this, (FlowsheetRow.__proto__ || Object.getPrototypeOf(FlowsheetRow)).apply(this, arguments));
    }

    _createClass(FlowsheetRow, [{
        key: "click",


        /**
         * clicks the flowsheet row
         * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
         * @param {{ shiftKey: boolean, ctrlKey: boolean }} opts - Click modifiers
         * @returns {Promise} - resolves when the action has completed
         */
        value: function click() {
            var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : { shiftKey: false, ctrlKey: false };

            /*
             * Note: We intentially forward the row click to the first cell in the row
             * because Flowsheet considers that to be the row-click operation.
             */
            return this.cell(0).click(opts);
        }

        /**
         * whether or not the cell is selected
         * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
         * @returns {Promise<boolean>} isSelected
         */

    }, {
        key: "isSelected",
        value: function isSelected() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("class");
            }).then(function (classNames) {
                return classNames.includes("is-selected");
            });
        }
    }, {
        key: "elementSelector",
        get: function get() {
            return ROOT_SELECTOR;
        }

        /**
         * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
         * @returns {List<module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row.cell>} cell
         */

    }, {
        key: "cell",
        get: function get() {
            return (0, _List2.default)({
                list: this,
                itemPage: _FlowsheetCell2.default,
                selectorList: this.selectorList,
                containerSelector: CELL_CONTAINER
            });
        }

        /**
         * the row data
         * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
         * @returns {Promise<object>} data
         */

    }, {
        key: "data",
        get: function get() {
            return this.cell.all("data").then(function (cells) {
                return { cells: cells };
            });
        }
    }]);

    return FlowsheetRow;
}(_UIComponent2.default);

exports.default = FlowsheetRow;
