// MPageFusion/testing/src/pages/controls/composite/form/controls/FormRadioButtonGroup.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _FormControl = require("../FormControl.page");

var _FormControl2 = _interopRequireDefault(_FormControl);

var _RadioButtonGroup = require("../../../composite/select-group/radio-button-group/RadioButtonGroup.page");

var _RadioButtonGroup2 = _interopRequireDefault(_RadioButtonGroup);

var _helpers = require("../helpers");

var _ClinicalBadge = require("../../../atomic/badge/ClinicalBadge.page");

var _ClinicalBadge2 = _interopRequireDefault(_ClinicalBadge);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-FormRadioButtonGroup";

/**
 * The FormRadioButtonGroupPage class
 * @class module:MPageFusion/testing.composite.form.FormRadioButtonGroupPage
 *
 * @example
 * const formRadioButtonGroup = new FormRadioButtonGroupPage(driver);
 * formRadiobuttonGroup.get("display")
 *     .then((display) => {
 *         assert(display === "Sample Label");
 *     });
 */

var FormRadioButtonGroupPage = function (_FormControlPage) {
    _inherits(FormRadioButtonGroupPage, _FormControlPage);

    function FormRadioButtonGroupPage(driver, options) {
        _classCallCheck(this, FormRadioButtonGroupPage);

        var _this = _possibleConstructorReturn(this, (FormRadioButtonGroupPage.__proto__ || Object.getPrototypeOf(FormRadioButtonGroupPage)).call(this, driver, options));

        _this._radiobuttonGroup = new _RadioButtonGroup2.default(driver, {
            selectorList: _this.selectorList.concat(_helpers.CLASSES.RIGHT_COLUMN)
        });
        return _this;
    }

    _createClass(FormRadioButtonGroupPage, [{
        key: "option",


        /**
         * clicks the radiobutton at the specified index
         * @memberOf module:MPageFusion/testing.composite.form.RadioButtonGroupPage
         * @param {Number} index - the index of the radiobutton to toggle
         * @returns {Promise} resolves when the action completes
         */
        value: function option(index) {
            var _this2 = this;

            return {
                click: function click() {
                    return _this2._radiobuttonGroup.option(index).click();
                }
            };
        }
    }, {
        key: "elementSelector",
        get: function get() {
            return ROOT_SELECTOR;
        }

        /**
         * Gets the current values in the radiobutton group
         * @memberOf module:MPageFusion/testing.composite.form.RadioButtonGroupPage
         * @returns {Promise<Array>} resolves with an array of objects { display }
         */

    }, {
        key: "value",
        get: function get() {
            return this._radiobuttonGroup.option.all().then(function (radiobuttonGroup) {
                return Promise.all(radiobuttonGroup.map(function (option) {
                    return Promise.all([option.get("display"), option.get("isSelected")]).then(function (_ref) {
                        var _ref2 = _slicedToArray(_ref, 2),
                            display = _ref2[0],
                            isSelected = _ref2[1];

                        return { display: display, isSelected: isSelected };
                    });
                }));
            });
        }
    }, {
        key: "isDisabled",
        get: function get() {
            return this._radiobuttonGroup.option.all().then(function (radiobuttonGroup) {
                return Promise.all(radiobuttonGroup.map(function (option) {
                    return Promise.all([option.get("isDisabled")]).then(function (_ref3) {
                        var _ref4 = _slicedToArray(_ref3, 1),
                            disabled = _ref4[0];

                        return { disabled: disabled };
                    });
                }));
            });
        }

        /**
         * returns the base RadionButtonGroup page
         * @memberOf module:MPageFusion/testing.composite.form.RadioButtonGroupPage
         * @returns {RadioButtonGroupPage} RadioButtonGroupPage
         */

    }, {
        key: "radiobuttonGroup",
        get: function get() {
            return this._radiobuttonGroup;
        }
    }]);

    return FormRadioButtonGroupPage;
}(_FormControl2.default);

exports.default = FormRadioButtonGroupPage;


_RadioButtonGroup2.default.CRITICALITY_TYPES = _ClinicalBadge2.default.TYPES;
