// MPageFusion/testing/src/pages/controls/composite/dialog/modal/ModalContent.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _UIComponent = require("../../../../UIComponent.page");

var _UIComponent2 = _interopRequireDefault(_UIComponent);

var _ModalHeader = require("./ModalHeader.page");

var _ModalHeader2 = _interopRequireDefault(_ModalHeader);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var CONTENT_SELECTOR = ".mpageui-Modal-content";
var BODY_SELECTOR = ".mpageui-Modal-body";
var FOOTER_SELECTOR = ".mpageui-Modal-footer";

var SIZE = {
    SMALL: "small",
    MEDIUM: "medium",
    LARGE: "large",
    CUSTOM: "custom"
};

/**
 * The ModalContentPage class
 * @alias module:MPageFusion/testing.composite.dialog.ModalPage.content
 * @class ModalContentPage
 * @private
 */

var ModalContentPage = function (_UIComponentPage) {
    _inherits(ModalContentPage, _UIComponentPage);

    function ModalContentPage() {
        _classCallCheck(this, ModalContentPage);

        return _possibleConstructorReturn(this, (ModalContentPage.__proto__ || Object.getPrototypeOf(ModalContentPage)).apply(this, arguments));
    }

    _createClass(ModalContentPage, [{
        key: "elementSelector",
        get: function get() {
            return CONTENT_SELECTOR;
        }
    }, {
        key: "header",
        get: function get() {
            return new _ModalHeader2.default(this.driver, {
                selectorList: this.selectorList
            });
        }
    }, {
        key: "bodySelectorList",
        get: function get() {
            return this.selectorList.concat(BODY_SELECTOR);
        }
    }, {
        key: "footerSelectorList",
        get: function get() {
            return this.selectorList.concat(FOOTER_SELECTOR);
        }
    }, {
        key: "size",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("class");
            }).then(function (classNames) {
                return Object.values(SIZE).find(function (style) {
                    return classNames.includes("mpageui-Modal--" + style);
                }) || "";
            });
        }
    }]);

    return ModalContentPage;
}(_UIComponent2.default);

exports.default = ModalContentPage;
