// MPageFusion/testing/src/pages/controls/atomic/nestable-pane/NestablePaneWithHeader.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _ContentHeader = _interopRequireDefault(require("../header/content-header/ContentHeader.page"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-NestablePane";
var VISIBLE = "mpageui-NestablePane-visible";
/**
 * The NestablePaneWithHeader class
 * @class module:MPageFusion/testing.atomic.panel.nestablePane.NestablePaneWithHeader
 *
 * @example
 * const nestablePane = new NestablePaneWithHeader(driver,"");
 * nestablePane.click()
 */

var NestablePaneWithHeader =
/*#__PURE__*/
function (_UIComponentPage) {
  _inheritsLoose(NestablePaneWithHeader, _UIComponentPage);

  function NestablePaneWithHeader(driver, args) {
    var _this;

    _this = _UIComponentPage.call(this, driver, args) || this;
    _this._contentHeader = new _ContentHeader["default"](driver, {
      selectorList: _this.selectorList,
      useExactSelector: true
    });
    return _this;
  }

  var _proto = NestablePaneWithHeader.prototype;

  /**
   * Closes the nestable pane
   * @memberOf module:MPageFusion/testing.atomic.panel.nestablePane.NestablePaneWithHeader
   * @returns {Promise<string>} click action
   */
  _proto.dismiss = function dismiss() {
    return this._contentHeader.leftAction.click();
  }
  /**
   * clicks the action
   * @memberOf module:MPageFusion/testing.atomic.panel.nestablePane.NestablePaneWithHeader
   * @returns {Promise} click action
   */
  ;

  _proto.click = function click() {
    return this._contentHeader.click();
  };

  _createClass(NestablePaneWithHeader, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * gets the nestablePane content selector list
     * @memberOf module:MPageFusion/testing.atomic.panel.nestablePane.NestablePaneWithHeader
     * @returns {Array} selector list
     */

  }, {
    key: "contentSelectorList",
    get: function get() {
      return this.selectorList;
    }
    /**
     * whether or not the nestablePane is visible
     * @memberOf module:MPageFusion/testing.atomic.panel.nestablePane.NestablePaneWithHeader
     * @returns {Promise<boolean>} isVisible
     */

  }, {
    key: "isVisible",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.getAttribute("class");
      }).then(function (classNames) {
        return classNames.includes(VISIBLE);
      });
    }
    /**
     * gets the nestable pane header display
     * @memberOf module:MPageFusion/testing.atomic.panel.nestablePane.NestablePaneWithHeader
     * @returns {Promise<string>} header display
     */

  }, {
    key: "headerDisplay",
    get: function get() {
      return this._contentHeader.display;
    }
  }]);

  return NestablePaneWithHeader;
}(_UIComponent["default"]);

exports["default"] = NestablePaneWithHeader;
