// gaia-plugin-injector/src/ui/gaia-plugin-injector/index.js
import * as Fusion from "MPageFusion";
import InjectorUI from "./InjectorUI";
import {MATCHING_MODES } from "./../../constants";
import { runCommand } from "Gaia";

const URLS_CHANGED = "InjectorIndex::UrlsChanged";
const SETTINGS_CHANGED = "InjectorIndex::SettingsChanged";
const MATCHING_MODE_CHANGED = "InjectorIndex::MatchingModeChanged";

const retrieveSettings = (control) =>
    runCommand("injector", "readSettings", {})
    .then(
        (settings) =>
            control
            .getChild("injectorUI")
            .setProp("settings", settings)
            .update()
    );

const retrieveMatchingMode = (control) =>
    runCommand("injector", "readMatchingMode")
    .then(matchingMode => 
        control
            .getChild("injectorUI")
            .setProp("matchingMode", matchingMode || MATCHING_MODES.STRICT)
            .update()
    );

const convertUrlsToUI = (urls) =>
    Object.keys(urls).reduce(
        (result, urlKey) =>
            result.concat({
                url: urlKey,
                top: urls[urlKey].top,
                bottom: urls[urlKey].bottom
            })
        , []
    );

const convertUrlsToConfig = (urls) =>
    urls.reduce(
        (result, url) => {
            result[url.url] = {
                top: url.top,
                bottom: url.bottom
            };
            return result;
        },
        {}
    );

const retrieveUrls = (control) =>
    runCommand("injector", "readUrls", {})
    .then(
        (urls) =>
            control
            .getChild("injectorUI")
            .setProp("urls", convertUrlsToUI(urls))
            .update()
    );

const saveSettings = (timeout, settings) => {
    if (timeout) {
        clearTimeout(timeout);
    }
    return setTimeout(
        () => runCommand("injector", "saveSettings", { settings }),
        200
    );
};

const changeMatchingMode = (timeout, matchingMode) => {
    if (timeout) {
        clearTimeout(timeout);
    }

    return setTimeout(
        () => runCommand("injector", "changeMatchingMode", { matchingMode }),
        200
    );
};

const saveUrls = (timeout, urls) => {
    if (timeout) {
        clearTimeout(timeout);
    }
    return setTimeout(
        () => runCommand("injector", "saveUrls", { urls: convertUrlsToConfig(urls) }),
        200
    );
};

class InjectorIndex extends Fusion.UIComponent {

    createChildren() {
        return [
            {
                injectorUI: new InjectorUI({
                    settingsChangedEventName: SETTINGS_CHANGED,
                    urlsChangedEventName: URLS_CHANGED,
                    matchingModeChangedEventName: MATCHING_MODE_CHANGED
                })
            },
        ];
    }

    afterCreate() {
        retrieveSettings(this);
        retrieveMatchingMode(this);
        retrieveUrls(this);

        this.on(URLS_CHANGED, (source, urls) => {
            this.getChild("injectorUI")
                .setProp("urls", urls)
                .update();
            this._saveUrlTimeout = saveUrls(
                this._saveUrlTimeout,
                urls
            );
        });

        this.on(SETTINGS_CHANGED, (source, settings) => {
            this.getChild("injectorUI")
                .setProp("settings", settings)
                .update();

            this._saveSettingsTimeout = saveSettings(
                this._saveSettingsTimeout,
                settings
            );
        });

        this.on(MATCHING_MODE_CHANGED, (source, matchingMode) => {
            this.getChild("injectorUI")
                .setProp("matchingMode", matchingMode)
                .update();

            this._changeMatchModeTimeout = changeMatchingMode(
                this._changeMatchModeTimeout,
                matchingMode
            );
        });
    }

    view(el, props, children, mChildren) {
        return  this.renderChildren();
    }
};


export default {
	label: "Injector",
	control: InjectorIndex,
    showHelp: true
};
