// gaia-plugin-contentserver/src/plugin.js
const commands = require("./commands");
const route = require("./route");

const gaiaJsonDescription =
      [
          "The full path of the file to read from. Defaults to the current",
          "directory's gaia.json if not provided. "
      ].join(" ");


module.exports = {
    id: "contentserver",
    commands: {
        addFolder: {
            description: "Adds a new folder to the content server",
            args: {
                alias: {
                    description: "Alias to access the folder from the browser"
                },
                path: {
                    description: "The full path to the folder"
                },
                gaiaJson: {
                    description: gaiaJsonDescription
                }

            },
            handler: commands.addFolder

        },
        removeFolder: {
            description: "Removes a folder to the content server",
            args: {
                alias: {
                    description: "Alias to access the folder from the browser"
                },
                gaiaJson: {
                    description: gaiaJsonDescription
                }
            },
            handler: commands.removeFolder
        },
        allFolders: {
            description: "Returns all folders in the content server",
            args: {
                gaiaJson: {
                    description: gaiaJsonDescription
                }
            },
            handler: commands.allFolders
        }
    },
    routes: {
        content: {
            description: [
                "Serves the content as specified by the addFolder command."
            ].join(""),
            handler: route.content
        }
    },
    ui: "dist/ui.js",
    markdown: {
        ui: "README.md"
    }
};
