// MPageFusion/testing/src/pages/controls/composite/combobox/Combobox.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _UIComponent = require("../../../UIComponent.page");

var _UIComponent2 = _interopRequireDefault(_UIComponent);

var _TextBox = require("../../atomic/textbox/TextBox.page");

var _TextBox2 = _interopRequireDefault(_TextBox);

var _Button = require("../../atomic/button/Button.page");

var _Button2 = _interopRequireDefault(_Button);

var _SearchListItem = require("../search/SearchListItem.page");

var _SearchListItem2 = _interopRequireDefault(_SearchListItem);

var _List = require("../../../helpers/List");

var _List2 = _interopRequireDefault(_List);

var _seleniumWebdriver = require("selenium-webdriver");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-ComboBox";
var DROP_DOWN_SELECTOR = ".mpageui-InlineDropDown";
var SEARCH_LIST_SELECTOR = ".mpageui-SearchList";

/**
 * The comboboxPage class
 * @class module:MPageFusion/test.composite.combobox.Comboboxpage
 *
 * @private
 */

var ComboboxPage = function (_UIComponentPage) {
    _inherits(ComboboxPage, _UIComponentPage);

    function ComboboxPage(driver, options) {
        _classCallCheck(this, ComboboxPage);

        var _this = _possibleConstructorReturn(this, (ComboboxPage.__proto__ || Object.getPrototypeOf(ComboboxPage)).call(this, driver, options));

        _this._button = new _Button2.default(driver, { selectorList: _this.selectorList });
        _this._textBox = new _TextBox2.default(driver, { selectorList: _this.selectorList });
        return _this;
    }

    _createClass(ComboboxPage, [{
        key: "focus",


        /**
         * focuses the textbox
         * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
         * @returns {Promise} resolves when the action is completed
         */
        value: function focus() {
            return this._textBox.focus();
        }

        /**
         * Whether the combobox is blur
         * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
         * @returns {Promise} resolves when the action is completed
         */

    }, {
        key: "blur",
        value: function blur() {
            return this._textBox.blur();
        }

        /**
         * enters text to the combobox
         * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
         * @param {String} text -the input text
         * @returns {Promise} resolves when the action is completed
         */

    }, {
        key: "enterText",
        value: function enterText(text) {
            return this._textBox.enterText(text);
        }

        /**
         * clears text from the combobox textbox
         * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
         * @returns {Promise} resolves when the action is completed
         */

    }, {
        key: "clearText",
        value: function clearText() {
            return this._textBox.clearText();
        }

        /**
         * clicks the dropdown icon button in the combobox bar
         * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
         * @returns {Promise} resolves when the action is completed
         */

    }, {
        key: "click",
        value: function click() {
            return this._button.click();
        }
    }, {
        key: "elementSelector",
        get: function get() {
            return ROOT_SELECTOR;
        }

        /**
         * whether the combobox bar is disabled
         * @memberOf module:MPageFusion/testing.composite.comboxbox.FormComboboxPage
         * @returns {Promise<boolean>} true or false
         */

    }, {
        key: "isDisabled",
        get: function get() {
            return Promise.all([this._textBox.isDisabled, this._button.isDisabled]).then(function (results) {
                return Boolean(results[0] && results[1]);
            });
        }

        /**
         * whether the search list is displayed
         * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
         * @returns {Promise<boolean>} true or false
         */

    }, {
        key: "isOpen",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.findElement(_seleniumWebdriver.By.css(DROP_DOWN_SELECTOR));
            }).then(function () {
                return true;
            }).catch(function () {
                return false;
            });
        }

        /**
         * returns placeholder text from the combobox
         * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
         * @returns {Promise<string>} placeholder text
         */

    }, {
        key: "placeholder",
        get: function get() {
            return this._textBox.placeholder;
        }

        /**
         * the value in the textbox
         * @memberOf module:MPageFusion/test.composite.combobox.Comboboxpage
         * @returns {Promise<String>} display text
         */

    }, {
        key: "value",
        get: function get() {
            return this._textBox.value;
        }

        /**
         * returns the item from the search list based on option(index)
         * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
         * @returns {List<module:MPageFusion/testing.composite.search.SearchPage.searchList>} searchList
         */

    }, {
        key: "option",
        get: function get() {
            return (0, _List2.default)({
                list: this,
                itemPage: _SearchListItem2.default,
                selectorList: this.selectorList,
                containerSelector: DROP_DOWN_SELECTOR + " " + SEARCH_LIST_SELECTOR
            });
        }

        /**
         * whether the combo box is focused
         * @memberOf module:MPageFusion/testing.composite.combobox.Comboboxpage
         * @returns {Promise<boolean>} true or false
         */

    }, {
        key: "isFocused",
        get: function get() {
            return this._textBox.isFocused;
        }
    }]);

    return ComboboxPage;
}(_UIComponent2.default);

exports.default = ComboboxPage;
