// gaia-plugin-milproxy/src/plugin.js
const commands = require("./commands");

module.exports = {
    id: "milproxy",
    commands: {
        start: {
            description: "Starts the millennium proxy server.",
            handler: commands.start,
            args: {}
        },
        stop: {
            description: "Stops the millennium proxy server.",
            handler: commands.stop,
            args: {}
        },
        status: {
            description: "Returns the status of the server",
            handler: commands.status,
            args: {}
        },
        login: {
            description: "Logins into a millennium domain",
            handler: commands.login,
            args: {
                username: {
                    description: "The millennium user"
                },
                password: {
                    description: "Password associated with the user"
                },
                domain: {
                    description: "The millennium domain"
                }
            }
        },
        readConfig: {
            description: "Reads config",
            handler: commands.readConfig
        },
        saveConfig: {
            description: "Saves config",
            handler: commands.saveConfig
        }
    },
    markdown: {
      ui: "README.md"
    },
    ui: "dist/ui.js"
};
