// MPageFusion/testing/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.NavigatorPage = exports.UIComponentPage = exports.helpers = exports.composite = exports.atomic = undefined;

var _UIComponent = require("./src/pages/UIComponent.page");

var _UIComponent2 = _interopRequireDefault(_UIComponent);

var _List = require("./src/pages/helpers/List");

var list = _interopRequireWildcard(_List);

var _AffirmDenyFooter = require("./src/pages/controls/atomic/footer/AffirmDenyFooter.page");

var _AffirmDenyFooter2 = _interopRequireDefault(_AffirmDenyFooter);

var _Anchor = require("./src/pages/controls/atomic/label/Anchor.page");

var _Anchor2 = _interopRequireDefault(_Anchor);

var _Button = require("./src/pages/controls/atomic/button/Button.page");

var _Button2 = _interopRequireDefault(_Button);

var _Card = require("./src/pages/controls/atomic/card/Card.page");

var _Card2 = _interopRequireDefault(_Card);

var _CardLayout = require("./src/pages/controls/atomic/grid/CardLayout.page");

var _CardLayout2 = _interopRequireDefault(_CardLayout);

var _CellGrid = require("./src/pages/controls/atomic/grid/CellGrid.page");

var _CellGrid2 = _interopRequireDefault(_CellGrid);

var _Checkbox = require("./src/pages/controls/atomic/checkbox/Checkbox.page");

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _ClinicalBadge = require("./src/pages/controls/atomic/badge/ClinicalBadge.page");

var _ClinicalBadge2 = _interopRequireDefault(_ClinicalBadge);

var _ContentDecorator = require("./src/pages/controls/atomic/content-decorator/ContentDecorator.page");

var _ContentDecorator2 = _interopRequireDefault(_ContentDecorator);

var _ContentHeader = require("./src/pages/controls/atomic/header/content-header/ContentHeader.page");

var _ContentHeader2 = _interopRequireDefault(_ContentHeader);

var _ContentPanel = require("./src/pages/controls/atomic/content-panel/ContentPanel.page");

var _ContentPanel2 = _interopRequireDefault(_ContentPanel);

var _DataSection = require("./src/pages/controls/atomic/data/DataSection.page");

var _DataSection2 = _interopRequireDefault(_DataSection);

var _DemographicsBanner = require("./src/pages/controls/atomic/demographics/DemographicsBanner.page");

var _DemographicsBanner2 = _interopRequireDefault(_DemographicsBanner);

var _Flowsheet = require("./src/pages/controls/atomic/flowsheet/Flowsheet.page");

var _Flowsheet2 = _interopRequireDefault(_Flowsheet);

var _FooterBase = require("./src/pages/controls/atomic/footer/FooterBase.page");

var _FooterBase2 = _interopRequireDefault(_FooterBase);

var _HideShow = require("./src/pages/controls/atomic/progressive-disclosure/HideShow.page");

var _HideShow2 = _interopRequireDefault(_HideShow);

var _index = require("./src/pages/controls/atomic/icon/index");

var icon = _interopRequireWildcard(_index);

var _Label = require("./src/pages/controls/atomic/label/Label.page");

var _Label2 = _interopRequireDefault(_Label);

var _LoadingSpinner = require("./src/pages/controls/atomic/indicator/LoadingSpinner.page");

var _LoadingSpinner2 = _interopRequireDefault(_LoadingSpinner);

var _NestablePane = require("./src/pages/controls/atomic/nestable-pane/NestablePane.page");

var _NestablePane2 = _interopRequireDefault(_NestablePane);

var _NestablePaneWithHeader = require("./src/pages/controls/atomic/nestable-pane/NestablePaneWithHeader.page");

var _NestablePaneWithHeader2 = _interopRequireDefault(_NestablePaneWithHeader);

var _NoResultsInfo = require("./src/pages/controls/atomic/indicator/NoResultsInfo.page");

var _NoResultsInfo2 = _interopRequireDefault(_NoResultsInfo);

var _NotificationFooter = require("./src/pages/controls/atomic/footer/NotificationFooter.page");

var _NotificationFooter2 = _interopRequireDefault(_NotificationFooter);

var _Overlay = require("./src/pages/controls/atomic/indicator/Overlay.page");

var _Overlay2 = _interopRequireDefault(_Overlay);

var _Select = require("./src/pages/controls/atomic/select/Select.Page");

var _Select2 = _interopRequireDefault(_Select);

var _Table = require("./src/pages/controls/atomic/table/Table.page");

var _Table2 = _interopRequireDefault(_Table);

var _TextBox = require("./src/pages/controls/atomic/textbox/TextBox.page");

var _TextBox2 = _interopRequireDefault(_TextBox);

var _Banner = require("./src/pages/controls/composite/banner/Banner.page");

var _Banner2 = _interopRequireDefault(_Banner);

var _BannerManager = require("./src/pages/controls/composite/banner/BannerManager.page");

var _BannerManager2 = _interopRequireDefault(_BannerManager);

var _ButtonGroup = require("./src/pages/controls/composite/button/ButtonGroup.page");

var _ButtonGroup2 = _interopRequireDefault(_ButtonGroup);

var _CollapseGroup = require("./src/pages/controls/composite/button/CollapseGroup/CollapseGroup.page");

var _CollapseGroup2 = _interopRequireDefault(_CollapseGroup);

var _ComponentHeader = require("./src/pages/controls/composite/component-header/ComponentHeader.page");

var _ComponentHeader2 = _interopRequireDefault(_ComponentHeader);

var _DetailPanel = require("./src/pages/controls/composite/panel/DetailPanel.page");

var _DetailPanel2 = _interopRequireDefault(_DetailPanel);

var _DetailPanelBanner = require("./src/pages/controls/composite/panel/DetailPanelBanner.page");

var _DetailPanelBanner2 = _interopRequireDefault(_DetailPanelBanner);

var _DetailPanelTitle = require("./src/pages/controls/composite/panel/DetailPanelTitle.page");

var _DetailPanelTitle2 = _interopRequireDefault(_DetailPanelTitle);

var _FormCheckbox = require("./src/pages/controls/composite/form/controls/FormCheckbox.page");

var _FormCheckbox2 = _interopRequireDefault(_FormCheckbox);

var _FormCheckboxGroup = require("./src/pages/controls/composite/form/controls/FormCheckboxGroup.page");

var _FormCheckboxGroup2 = _interopRequireDefault(_FormCheckboxGroup);

var _FormCombobox = require("./src/pages/controls/composite/form/controls/FormCombobox.page");

var _FormCombobox2 = _interopRequireDefault(_FormCombobox);

var _FormRadioButtonGroup = require("./src/pages/controls/composite/form/controls/FormRadioButtonGroup.page");

var _FormRadioButtonGroup2 = _interopRequireDefault(_FormRadioButtonGroup);

var _FormSearch = require("./src/pages/controls/composite/form/controls/FormSearch.page");

var _FormSearch2 = _interopRequireDefault(_FormSearch);

var _FormSelect = require("./src/pages/controls/composite/form/controls/FormSelect.page");

var _FormSelect2 = _interopRequireDefault(_FormSelect);

var _FormTextBox = require("./src/pages/controls/composite/form/controls/FormTextBox.page");

var _FormTextBox2 = _interopRequireDefault(_FormTextBox);

var _FormTextArea = require("./src/pages/controls/composite/form/controls/FormTextArea.page");

var _FormTextArea2 = _interopRequireDefault(_FormTextArea);

var _InlineList = require("./src/pages/controls/composite/list/InlineList.Page");

var _InlineList2 = _interopRequireDefault(_InlineList);

var _Menu = require("./src/pages/controls/composite/menu/Menu.page");

var _Menu2 = _interopRequireDefault(_Menu);

var _Modal = require("./src/pages/controls/composite/dialog/modal/Modal.page");

var _Modal2 = _interopRequireDefault(_Modal);

var _MenuButton = require("./src/pages/controls/composite/button/MenuButton.page");

var _MenuButton2 = _interopRequireDefault(_MenuButton);

var _Popup = require("./src/pages/controls/composite/dialog/popup/Popup.page");

var _Popup2 = _interopRequireDefault(_Popup);

var _Search = require("./src/pages/controls/composite/search/Search.page");

var _Search2 = _interopRequireDefault(_Search);

var _Navigator = require("./src/pages/navigator/Navigator.page");

var _Navigator2 = _interopRequireDefault(_Navigator);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @alias module:MPageFusion/testing.atomic
 */


// atomic
/** @module MPageFusion/testing */

var atomic = {
    /** @alias module:MPageFusion/testing.atomic.badge */
    badge: {
        ClinicalBadgePage: _ClinicalBadge2.default
    },
    /** @alias module:MPageFusion/testing.atomic.button */
    button: {
        ButtonPage: _Button2.default
    },
    /** @alias module:MPageFusion/testing.atomic.card */
    card: {
        CardPage: _Card2.default
    },
    /** @alias module:MPageFusion/testing.atomic.checkbox */
    checkbox: {
        CheckboxPage: _Checkbox2.default
    },
    /** @alias module:MPageFusion/testing.atomic.demographics */
    demographics: {
        DemographicsBannerPage: _DemographicsBanner2.default
    },
    /** @alias module:MPageFusion/testing.atomic.flowsheet */
    flowsheet: {
        FlowsheetPage: _Flowsheet2.default
    },
    /** @alias module:MPageFusion/testing.atomic.footer */
    footer: {
        AffirmDenyFooterPage: _AffirmDenyFooter2.default,
        NotificationFooterPage: _NotificationFooter2.default,
        FooterBasePage: _FooterBase2.default
    },
    /** @alias module:MPageFusion/testing.atomic.grid */
    grid: {
        CardLayoutPage: _CardLayout2.default,
        CellGridPage: _CellGrid2.default
    },
    /** @alias module:MPageFusion/testing.atomic.header */
    header: {
        ContentHeaderPage: _ContentHeader2.default
    },
    /** @alias module:MPageFusion/testing.atomic.icon */
    icon: icon,
    /** @alias module:MPageFusion/testing.atomic.indicator */
    indicator: {
        NoResultsInfoPage: _NoResultsInfo2.default,
        OverlayPage: _Overlay2.default,
        LoadingSpinnerPage: _LoadingSpinner2.default
    },
    /** @alias module:MPageFusion/testing.atomic.panel */
    panel: {
        nestablePane: {
            NestablePanePage: _NestablePane2.default,
            NestablePaneWithHeaderPage: _NestablePaneWithHeader2.default
        }
    },
    /** @alias module:MPageFusion/testing.atomic.progressiveDisclosure */
    progressiveDisclosure: {
        HideShowPage: _HideShow2.default
    },
    /** @alias module:MPageFusion/testing.atomic.select */
    select: {
        SelectPage: _Select2.default
    },
    /** @alias module:MPageFusion/testing.atomic.table */
    table: {
        TablePage: _Table2.default
    },
    /** @alias module:MPageFusion/testing.atomic.textbox */
    textbox: {
        TextBoxPage: _TextBox2.default
    },
    /** @alias module:MPageFusion/testing.atomic.label */
    label: {
        LabelPage: _Label2.default,
        AnchorPage: _Anchor2.default
    },
    /** @alias module:MPageFusion/testing.atomic.contentPanel */
    contentPanel: {
        ContentPanelPage: _ContentPanel2.default
    },
    /** @alias module:MPageFusion/testing.atomic.decorator */
    decorator: {
        ContentDecoratorPage: _ContentDecorator2.default
    },
    /** @alias module:MPageFusion/testing.atomic.data */
    data: {
        DataSectionPage: _DataSection2.default
    }
};

/**
 * @alias module:MPageFusion/testing.composite
 */


// composite
var composite = {
    /** @alias module:MPageFusion/testing.composite.banner */
    banner: {
        BannerPage: _Banner2.default,
        BannerManagerPage: _BannerManager2.default
    },
    /** @alias module:MPageFusion/testing.composite.button */
    button: {
        ButtonGroupPage: _ButtonGroup2.default,
        CollapseGroupPage: _CollapseGroup2.default,
        MenuButtonPage: _MenuButton2.default
    },
    /** @alias module:MPageFusion/testing.composite.dialog */
    dialog: {
        PopupPage: _Popup2.default,
        ModalPage: _Modal2.default
    },
    ComponentHeaderPage: _ComponentHeader2.default,
    /** @alias module:MPageFusion/testing.composite.form */
    form: {
        FormCheckboxPage: _FormCheckbox2.default,
        FormCheckboxGroupPage: _FormCheckboxGroup2.default,
        FormComboboxPage: _FormCombobox2.default,
        FormRadioButtonGroupPage: _FormRadioButtonGroup2.default,
        FormSearchPage: _FormSearch2.default,
        FormSelectPage: _FormSelect2.default,
        FormTextBoxPage: _FormTextBox2.default,
        FormTextAreaPage: _FormTextArea2.default
    },
    /** @alias module:MPageFusion/testing.composite.list */
    list: {
        InlineListPage: _InlineList2.default
    },
    /** @alias module:MPageFusion/testing.composite.menu */
    menu: {
        MenuPage: _Menu2.default
    },
    /** @alias module:MPageFusion/testing.composite.search */
    search: {
        SearchPage: _Search2.default
    },
    /** @alias module:MPageFusion/testing.composite.panel */
    panel: {
        DetailPanelPage: _DetailPanel2.default,
        DetailPanelBannerPage: _DetailPanelBanner2.default,
        DetailPanelTitlePage: _DetailPanelTitle2.default
    }
};

/**
 * @alias module:MPageFusion/testing.helpers
 */
var helpers = {
    list: list
};

exports.atomic = atomic;
exports.composite = composite;
exports.helpers = helpers;
exports.UIComponentPage = _UIComponent2.default;
exports.NavigatorPage = _Navigator2.default;
