// gaia-plugin-linter/src/plugin.js
const commands = require("./commands");

module.exports = {
    id: "linter",
    commands: {
        run: {
            description: "Runs ESLint on files in src/main",
            handler: commands.run,
            args: {
                srcDir: "Set the path to js files. Start at package.json level. Default is 'src/main'.",
                webFlag: "Indicates if the command was ran from console or web",
                fix: "Indicates if eslint needs to be run in fix mode, by default it is set to false"
            }
        },
        cli: {
            description: "Runs ESLint CLI on files in src/main",
            handler: commands.runCLI,
            args: {
                maxWarnings: "The maximum number of lint warnings allowed -- unset by default",
                srcDir: "The path of the files to be linted, default is 'src/main'",
                fix: "Indicates if eslint needs to be run in fix mode, by default it is set to false"
            }
        },
        determineSrcPath: {
            description: "Checks if there is a settings for the plugin in the gaia.json config file",
            handler: commands.determineSrcPath,
            args: {}
        },
        saveSrcPath: {
            description: "Writes the JSON configuration for the path of files to be linted",
            handler: commands.saveSrcPath,
            args: {
                srcDir: "The path of the files to be linted, default is 'src/main'"
            }
        }
    },
    markdown: {
        ui: "docs/ui.md"
    },
    ui: "dist/ui.js"
};
