// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/indicator/LoadingSpinner.js
"use strict";
import UIComponent from "../../../base";
import classNames from "classnames";

/**
 * Spinner Sizes
 * @type {{SMALL: string, MEDIUM: string, LARGE: string, XSMALL: string}}
 */
const SIZE = {
    SMALL: "mpageui-LoadingSpinner-icon--sm",
    MEDIUM: "mpageui-LoadingSpinner-icon--md",
    LARGE: "mpageui-LoadingSpinner-icon--lg",
    XSMALL: "mpageui-LoadingSpinner-icon--xs"
};

/**
 * Maps the loading spinner sizes to the appropriate legacy image class
 * so we can use the spinner image.
 * @param {string} size - The loading spinner size.
 * @returns {string} The legacy loading spinner class image.
 */
const legacyImage = (size) => ({
    [ SIZE.XSMALL ]: "image-6439_12",
    [ SIZE.SMALL ]: "image-6439_16",
    [ SIZE.MEDIUM ]: "image-6439_24",
    [ SIZE.LARGE ]: "image-6439_32"
})[ size ];

/**
 * UIComponent shows the loading spinner when needed
 * @class LoadingSpinner
 */
class LoadingSpinner extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            classNames: "",
            isLoading: false,
            size: SIZE.MEDIUM
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props, children) {
        return props.isLoading ?
            el(
                "div",
                {
                    class: classNames({
                        [ props.classNames ]: props.classNames.length > 0,
                        [ props.size ]: true,
                        [ legacyImage(props.size) ]: true
                    })
                }
            ) : [];
    }
}

LoadingSpinner.SIZE = SIZE;
export default LoadingSpinner;
