// MPageFusion/testing/src/pages/controls/composite/component-header/ComponentHeader.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _seleniumWebdriver = require("selenium-webdriver");

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _Button = _interopRequireDefault(require("../../atomic/button/Button.page"));

var _ButtonGroup = _interopRequireDefault(require("../../composite/button/ButtonGroup.page"));

var _Popup = _interopRequireDefault(require("../dialog/popup/Popup.page"));

var _Menu = _interopRequireDefault(require("../menu/Menu.page"));

var _Label = _interopRequireDefault(require("../../atomic/label/Label.page"));

var _ResponsiveButtonGroup = _interopRequireDefault(require("../../composite/button/ResponsiveButtonGroup.page"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-ComponentHeader";
var HEADER_TITLE = ".mpageui-ComponentHeader-title";
var COUNT = ".mpageui-ComponentHeader-count";
var SUB_LABEL = ".mpageui-ComponentHeader-sublabel";
var LOOKBACK_CLASS = "mpageui-ComponentHeader";
var ERROR_ICON = ".mpageui-ComponentHeader-errorIcon";
var GAPCHECK_ICON = ".mpageui-ComponentHeader-gapcheckicon";
var SPLIT_BUTTON = ".mpageui-SplitButton-button";
var PLUS_ADD = ".mpageui-ComponentHeader-plusadd";
var HEADER_REFRESH = ".mpageui-ComponentHeader-refresh";
var BUTTON_GROUP = ".mpageui-ButtonGroup";
var HEADER_TOGGLES = ".mpageui-ComponentHeader-toggles";
var SVG_MENU = ".mpageui-SvgIcon.mpageui-SvgIcon-Menu";
var OVERLAY_SELECTOR = ".mpageui-Popup-layover";
var TYPES = {
  SUBHEADER: "lookbackSubheader",
  DEFAULT: "defaultLookback",
  LOOKBACK_LABEL: "lookbackLabel"
};
var LOOKBACK_SELECTOR = ".mpageui-ComponentHeader-lookbackSelector";
/**
 * The ComponentHeaderPage class
 * @class module:MPageFusion/testing.composite.ComponentHeaderPage
 *
 * @example
 * const componentHeader = new ComponentHeaderPage(driver);
 * componentHeader.toggleButtons.button().display;
 * componentHeader.refreshButton.click();
 */

var ComponentHeaderPage =
/*#__PURE__*/
function (_UIComponentPage) {
  _inheritsLoose(ComponentHeaderPage, _UIComponentPage);

  function ComponentHeaderPage() {
    return _UIComponentPage.apply(this, arguments) || this;
  }

  _createClass(ComponentHeaderPage, [{
    key: "elementSelector",

    /**
     * @private
     */
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
    * gets the title of component header
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<String>} title
    */

  }, {
    key: "title",
    get: function get() {
      return this.getNestedElement(HEADER_TITLE).then(function (element) {
        return element.getText();
      });
    }
    /**
    * checks whether the title is shown as link
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<Boolean>} isTitleShownAsLink
    */

  }, {
    key: "isTitleShownAsLink",
    get: function get() {
      return Promise.resolve(this.getNestedElement(HEADER_TITLE).then(function (element) {
        return element.getTagName();
      })).then(function (element) {
        return element === "a";
      });
    }
    /**
    * gets the count
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<String>} count
    */

  }, {
    key: "count",
    get: function get() {
      return this.getNestedElement(COUNT).then(function (element) {
        return element.getText();
      });
    }
    /**
    * gets sub label in the component header
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<String>} subLabel
    */

  }, {
    key: "subLabel",
    get: function get() {
      return this.getNestedElement(SUB_LABEL).then(function (element) {
        return element.getText();
      });
    }
    /**
    * gets lookbackLabel in the component header
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<String>} lookbackLabel
    */

  }, {
    key: "lookbackLabel",
    get: function get() {
      var _this = this;

      return this.getNestedElement(new _Label["default"](this.driver).elementSelector).then(function (element) {
        return element.getAttribute("class");
      }).then(function (classNames) {
        return Object.values(TYPES).find(function (type) {
          return classNames.includes(LOOKBACK_CLASS + "-" + type) || "";
        });
      }).then(function (css) {
        return _this.getNestedElement("." + LOOKBACK_CLASS + "-" + css);
      }).then(function (element) {
        return element.getText();
      });
    }
    /**
    * checks whether error icon is shown.
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<Boolean>} isErrorIconShown
    */

  }, {
    key: "isErrorIconShown",
    get: function get() {
      return this.getNestedElement(ERROR_ICON).then(function (element) {
        return Boolean(element);
      });
    }
    /**
    * checks whether gap check indicator is shown
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<Boolean>} isGapCheckIndicatorShown
    */

  }, {
    key: "isGapCheckIndicatorShown",
    get: function get() {
      return this.getNestedElement(GAPCHECK_ICON).then(function (element) {
        return Boolean(element);
      });
    }
    /**
    * checks whether plusadd button is split
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<Boolean>} isPlusAddSplitButton
    */

  }, {
    key: "isPlusAddSplitButton",
    get: function get() {
      return this.getNestedElement(SPLIT_BUTTON).then(function (element) {
        return Boolean(element);
      });
    }
    /**
    * gets the plusadd button
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {module:MPageFusion/testing.atomic.button.Button} plusAddButton
    */

  }, {
    key: "plusAddButton",
    get: function get() {
      return new _Button["default"](this.driver, {
        selectorList: this.selectorList.concat(PLUS_ADD),
        useExactSelector: true
      });
    }
    /**
    * gets the plusadd popup
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {module:MPageFusion/testing.composite.dialog.popup.PopupPage} plusAddPopup
    */

  }, {
    key: "plusAddPopup",
    get: function get() {
      return new _Popup["default"](this.driver, {
        selectorList: this.selectorList.concat(PLUS_ADD)
      });
    }
    /**
    * checks whether add button is shown
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<Boolean>} isAddButtonShown
    */

  }, {
    key: "isAddButtonShown",
    get: function get() {
      return this.getNestedElement(PLUS_ADD).then(function (element) {
        return Boolean(element);
      });
    }
    /**
    * checks whether refresh button is shown
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise<Boolean>} isRefreshButtonShown
    */

  }, {
    key: "isRefreshButtonShown",
    get: function get() {
      return this.getNestedElement(HEADER_REFRESH).then(function (element) {
        return Boolean(element);
      });
    }
    /**
    * gets the refresh button
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {module:MPageFusion/testing.atomic.button.Button} refreshButton
    */

  }, {
    key: "refreshButton",
    get: function get() {
      return new _Button["default"](this.driver, {
        selectorList: this.selectorList.concat(HEADER_REFRESH),
        useExactSelector: true
      });
    }
    /**
    * gets the toggle button group
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {module:MPageFusion/testing.composite.button.ButtonGroup} toggleButtons
    */

  }, {
    key: "toggleButtons",
    get: function get() {
      return new _ButtonGroup["default"](this.driver, {
        selectorList: this.selectorList.concat("" + BUTTON_GROUP + HEADER_TOGGLES),
        useExactSelector: true
      });
    }
    /**
    * gets the component menu
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {module:MPageFusion/testing.composite.menu.Menu} componentMenu
    */

  }, {
    key: "componentMenu",
    get: function get() {
      return new _Menu["default"](this.driver);
    }
    /**
    * opens the component menu
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {module:MPageFusion/testing.composite.button.ButtonGroup} componentMenuOpen
    * @returns {Promise} resolves when the action completes
    */

  }, {
    key: "componentMenuOpen",
    get: function get() {
      return this.getNestedElement(SVG_MENU).then(function (element) {
        return element.click();
      });
    }
    /**
    * closes the component menu
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {Promise} resolves when the action completes
    */

  }, {
    key: "componentMenuClose",
    get: function get() {
      return this.driver.findElement(_seleniumWebdriver.By.css(OVERLAY_SELECTOR)).then(function (element) {
        return element.click();
      });
    }
    /**
    * This method gets the ResponsiveButtonGroup page object.
    * @memberOf module:MPageFusion/testing.composite.ComponentHeaderPage
    * @returns {module:MPageFusion/testing.composite.button.responsiveButtonGroup} lookbackOptions
    */

  }, {
    key: "lookbackOptions",
    get: function get() {
      return new _ResponsiveButtonGroup["default"](this.driver, {
        selectorList: this.selectorList.concat(LOOKBACK_SELECTOR)
      });
    }
  }]);

  return ComponentHeaderPage;
}(_UIComponent["default"]);

exports["default"] = ComponentHeaderPage;
