// MPageFusion/testing/src/pages/controls/atomic/grid/CardLayout.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _UIComponent = require("../../../UIComponent.page");

var _UIComponent2 = _interopRequireDefault(_UIComponent);

var _List = require("../../../helpers/List");

var _List2 = _interopRequireDefault(_List);

var _CardCell = require("./CardCell.page");

var _CardCell2 = _interopRequireDefault(_CardCell);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-CellGrid";
var ROW_SELECTOR = ".mpageui-CellGrid-row";

/**
 * The CardLayoutPage class
 * @class module:MPageFusion/testing.atomic.grid.CardLayoutPage
 *
 * @example
 * const cardLayout = new CardLayoutPage(driver);
 */

var CardLayoutPage = function (_UIComponentPage) {
    _inherits(CardLayoutPage, _UIComponentPage);

    function CardLayoutPage() {
        _classCallCheck(this, CardLayoutPage);

        return _possibleConstructorReturn(this, (CardLayoutPage.__proto__ || Object.getPrototypeOf(CardLayoutPage)).apply(this, arguments));
    }

    _createClass(CardLayoutPage, [{
        key: "elementSelector",
        get: function get() {
            return ROOT_SELECTOR;
        }

        /**
         * @memberOf module:MPageFusion/testing.atomic.grid.CardLayoutPage
         * @returns {List<module:MPageFusion/testing.atomic.grid.CardCellPage>} returns list of children
         */

    }, {
        key: "cellDetails",
        get: function get() {
            return (0, _List2.default)({
                list: this,
                itemPage: _CardCell2.default,
                selectorList: this.selectorList,
                containerSelector: ROW_SELECTOR
            });
        }
    }]);

    return CardLayoutPage;
}(_UIComponent2.default);

exports.default = CardLayoutPage;
