// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/list/List.js
import UIComponent from "../../../base";
import NestablePaneRoot from "../../atomic/nestable-pane/NestablePaneRoot";
import classNames from "classnames";

/**
 * A vertical list of items. All the allowed items are prefixed with "List"
 * on the name of their classes. Nested lists can be achieved by adding instances
 * of SubList.
 */
export default class List extends UIComponent {

    /**
     * @inheritdoc
     */
    initialProps() {
        return {
            classNames: "",
            autoInsertDividers: true,
            defaultClassNames: "mpageui-List",
            content: [],
            isDecorated: false
        };
    }

    /**
     * @inheritdoc
     */
    createChildren() {
        return [
            { nestablePane: new NestablePaneRoot() }
        ];
    }

    /**
     * @inheritdoc
     */
    propChangeHandlers() {
        return {
            content: (content) => this.getChild("nestablePane").replaceAllChildren(content),
            classNames: (classes) => {
                this.getChild("nestablePane")
                    .setProp("classNames", classNames({
                        [ classes ]: classes.length > 0,
                        [ this.getProp("defaultClassNames") ]: this.getProp("defaultClassNames").length > 0,
                        "mpageui-List--autodividers": this.getProp("autoInsertDividers"),
                        "mpageui-List--decorated": this.getProp("isDecorated")
                    }));
            }
        };
    }

    /**
     * @inheritdoc
     */
    view(el, props, children, mChildren) {
        return this.renderChildren();
    }
}
