// gaia-plugin-unit-testing/src/ui/js/ResultsSummary.js
import { UIComponent } from "MPageFusion";
import i18n from "../../i18n/en_us";
import "../../styles/UnitTester.less";

/**
 * Displays the summary of the unit tests results for a browser.
 * @class ResultsSummary
 */
export default class ResultsSummary extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            summary: {
                success: "",
                failed: "",
                skipped: "",
                error: false
            }
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props) {
        return el(
            "div",
            { class: "mpageui-u-pad-bottom-relative-base" },
            [
                el("div", { class: "results-success" }, `${i18n.content.resultsSummary.success} ${props.summary.success}`),
                el("div", { class: "results-failed" }, `${i18n.content.resultsSummary.failed} ${props.summary.failed}`),
                el("div", { class: "results-skipped" }, `${i18n.content.resultsSummary.skipped} ${props.summary.skipped}`),
                el("div", {}, `${i18n.content.resultsSummary.error} ${props.summary.error}`)
            ]

        )
    }
}
