// MPageFusion/testing/src/pages/controls/atomic/button/Button.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _UIComponent = require("../../../UIComponent.page");

var _UIComponent2 = _interopRequireDefault(_UIComponent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TYPES = {
    PRIMARY: "primary",
    SECONDARY: "secondary",
    LINK: "link",
    DEFAULT: ""
};

var ROOT_SELECTOR = ".mpageui-Button";

/**
 * The ButtonPage class
 * @class module:MPageFusion/testing.atomic.button.ButtonPage
 *
 * @example
 * const button = new ButtonPage(driver);
 */

var ButtonPage = function (_UIComponentPage) {
    _inherits(ButtonPage, _UIComponentPage);

    function ButtonPage() {
        _classCallCheck(this, ButtonPage);

        return _possibleConstructorReturn(this, (ButtonPage.__proto__ || Object.getPrototypeOf(ButtonPage)).apply(this, arguments));
    }

    _createClass(ButtonPage, [{
        key: "click",


        /**
         * clicks the button
         * @memberOf module:MPageFusion/testing.atomic.button.ButtonPage
         * @returns {Promise} resolves when the action is completed
         */
        value: function click() {
            return this.getRootElement().then(function (element) {
                return element.click();
            });
        }
    }, {
        key: "elementSelector",
        get: function get() {
            return ROOT_SELECTOR;
        }

        /**
         * the selector list for the button icon
         * @memberOf module:MPageFusion/testing.atomic.button.ButtonPage
         * @returns {Array} selector
         */

    }, {
        key: "iconSelectorList",
        get: function get() {
            return this.selectorList;
        }

        /**
         * the button display
         * @memberOf module:MPageFusion/testing.atomic.button.ButtonPage
         * @returns {Promise<string>} display
         */

    }, {
        key: "display",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getText();
            });
        }

        /**
         * the button type
         * @memberOf module:MPageFusion/testing.atomic.button.ButtonPage
         * @returns {Promise<string>} type
         */

    }, {
        key: "type",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("class");
            }).then(function (classNames) {
                return Object.values(TYPES).filter(function (type) {
                    return type !== TYPES.DEFAULT;
                }).find(function (type) {
                    return classNames.includes("mpageui-Button--" + type);
                }) || TYPES.DEFAULT;
            });
        }

        /**
         * whether the button is disabled
         * @memberOf module:MPageFusion/testing.atomic.button.ButtonPage
         * @returns {Promise<boolean>} isDisabled
         */

    }, {
        key: "isDisabled",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("disabled");
            }).then(function (attribute) {
                return Boolean(attribute);
            });
        }

        /**
         * whether the button is selected
         * @memberOf module:MPageFusion/testing.atomic.button.ButtonPage
         * @returns {Promise<boolean>} isSelected
         */

    }, {
        key: "isSelected",
        get: function get() {
            return this.getRootElement().then(function (element) {
                return element.getAttribute("aria-selected");
            }).then(function (attribute) {
                return Boolean(attribute);
            });
        }
    }]);

    return ButtonPage;
}(_UIComponent2.default);

exports.default = ButtonPage;


ButtonPage.TYPES = TYPES;
