// gaia-plugin-unit-testing/src/ui/js/Body.js
import { composite, UIComponent } from "MPageFusion";
import i18n from "../../i18n/en_us";
import UnitTester, { runAfterConfigSave } from "./UnitTester";
import Coverage from "./Coverage";
import Config from "./Config";

const RUN_KARMA = "Config::runKarma";
const {
    navigation: {
        TabControl
        }
    } = composite;

/**
 * This class creates the three tabs that make up the JS Unit Testing tab in Gaia: Unit Tester, Coverage, and Config
 * @class Body
 */
export default class Body extends UIComponent {

    /**
     * @inheritdoc
     */
    createChildren() {
        return [
            {
                tabs: new TabControl({
                    content: [{
                        id: "tests",
                        display: i18n.tabs.tests,
                        view: new UnitTester()
                    }, {
                        id: "coverage",
                        display: i18n.tabs.coverage,
                        view: new Coverage()
                    }, {
                        id: "config",
                        display: i18n.tabs.config,
                        view: new Config({
                            runKarmaEventName: RUN_KARMA
                        })
                    }]
                })
            }
        ];
    }

    /**
     * @inheritdoc
     */
    afterCreate() {
        //The Config tab will emit this event when it has saved the config and wants the tests ran
        this.on(RUN_KARMA, () => {
            this.stopPropagation(RUN_KARMA);
            const testerTab = this.getChild("tabs").getProp("content").find(tab => tab.id === "tests").view;
            runAfterConfigSave(testerTab);
        });
    }

    /**
     * @inheritdoc
     */
    view(el, props, children, mappedChildren) {
        return mappedChildren.tabs.render();
    }
};
