// MPageFusion/testing/src/pages/controls/atomic/flowsheet/FlowsheetCell.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _click = require("../../../helpers/click");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-Flowsheet-row-cell";
/**
 * The FlowsheetCellPage class
 * @alias module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row.cell
 * @class FlowsheetCellPage
 * @private
 */

var FlowsheetCell =
/*#__PURE__*/
function (_UIComponentPage) {
  _inheritsLoose(FlowsheetCell, _UIComponentPage);

  function FlowsheetCell() {
    return _UIComponentPage.apply(this, arguments) || this;
  }

  var _proto = FlowsheetCell.prototype;

  /**
   * clicks the flowsheet cell
   * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row.cell
   * @param {{ shiftKey: boolean, ctrlKey: boolean }} opts - Click modifiers
   * @returns {Promise} - resolves when the action has completed
   */
  _proto.click = function click(opts) {
    if (opts === void 0) {
      opts = {
        shiftKey: false,
        ctrlKey: false
      };
    }

    return this.getRootElement().then((0, _click.clickModifier)(this.driver)(opts));
  }
  /**
   * the cell display
   * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row.cell
   * @returns {Promise<string>} display
   */
  ;

  _createClass(FlowsheetCell, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
  }, {
    key: "contentSelectorList",
    get: function get() {
      return this.selectorList;
    }
  }, {
    key: "display",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.getText();
      });
    }
    /**
     * the cell data
     * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row.cell
     * @returns {Promise<Object>} data
     */

  }, {
    key: "data",
    get: function get() {
      return this.get("display").then(function (display) {
        return {
          display: display
        };
      });
    }
  }]);

  return FlowsheetCell;
}(_UIComponent["default"]);

exports["default"] = FlowsheetCell;
