// MPageFusion/testing/src/pages/controls/atomic/flowsheet/FlowsheetRow.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _FlowsheetCell = _interopRequireDefault(require("./FlowsheetCell.page"));

var _List = _interopRequireDefault(require("../../../helpers/List"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var ROOT_SELECTOR = ".mpageui-Flowsheet-row-wrapper";
var CELL_CONTAINER = ".mpageui-Flowsheet-row";
/**
 * The FlowsheetRowPage class
 * @alias module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
 * @class FlowsheetRowPage
 * @private
 */

var FlowsheetRow =
/*#__PURE__*/
function (_UIComponentPage) {
  _inheritsLoose(FlowsheetRow, _UIComponentPage);

  function FlowsheetRow() {
    return _UIComponentPage.apply(this, arguments) || this;
  }

  var _proto = FlowsheetRow.prototype;

  /**
   * clicks the flowsheet row
   * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
   * @param {{ shiftKey: boolean, ctrlKey: boolean }} opts - Click modifiers
   * @returns {Promise} - resolves when the action has completed
   */
  _proto.click = function click(opts) {
    if (opts === void 0) {
      opts = {
        shiftKey: false,
        ctrlKey: false
      };
    }

    /*
     * Note: We intentially forward the row click to the first cell in the row
     * because Flowsheet considers that to be the row-click operation.
     */
    return this.cell(0).click(opts);
  }
  /**
   * whether or not the cell is selected
   * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
   * @returns {Promise<boolean>} isSelected
   */
  ;

  _proto.isSelected = function isSelected() {
    return this.getRootElement().then(function (element) {
      return element.getAttribute("class");
    }).then(function (classNames) {
      return classNames.includes("is-selected");
    });
  };

  _createClass(FlowsheetRow, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
     * @returns {List<module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row.cell>} cell
     */

  }, {
    key: "cell",
    get: function get() {
      return (0, _List["default"])({
        list: this,
        itemPage: _FlowsheetCell["default"],
        selectorList: this.selectorList,
        containerSelector: CELL_CONTAINER
      });
    }
    /**
     * the row data
     * @memberOf module:MPageFusion/testing.atomic.flowsheet.FlowsheetPage.row
     * @returns {Promise<object>} data
     */

  }, {
    key: "data",
    get: function get() {
      return this.cell.all("data").then(function (cells) {
        return {
          cells: cells
        };
      });
    }
  }]);

  return FlowsheetRow;
}(_UIComponent["default"]);

exports["default"] = FlowsheetRow;
