// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/tabs/TabContents.js
"use strict";
import UIComponent from "../../../base";
import classNames from "classnames";

const NAMESPACE = "mpageui-TabContents";

export default class TabContents extends UIComponent {
    constructor(props, children) {
        super(props, children);
    }

    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            defaultClassNames: NAMESPACE,
            classNames: "",
            panels: [],
            selectedTabId: ""
        };
    }

    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return {
            panels: (panels) => {
                this.replaceAllChildren(panels);
            },
            selectedTabId: (selectedTabId) => {
                this.getProp("panels").forEach((p) => {
                    p.setProp("isSelected", p.getProp("tabLink") === selectedTabId);
                });
            }
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mChildren) {
        return el(
            "div",
            {
                "class": classNames({
                    [ props.defaultClassNames ]: true,
                    [ props.classNames ]: (props.classNames !== "")
                }),
                role: "tabpanel"
            },
            this.renderChildren()
        );
    }
}
