// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/content-panel/ContentPanelBody.js
"use strict";

import UIComponent from "../../../base";
import classNames from "classnames";

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------
const NAMESPACE = "mpageui-ContentPanel-body";

/**
 * The ContentPanelBody class.
 * @class ContentPanelBody
 */
export default class ContentPanelBody extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            bodyBgStyle: "",
            bodyClassNames: "",
            content: []
        };
    }

    /**
     * @inheritDoc
     */
    propChangeHandlers() {
        return {
            content: (content = []) => {
                this.replaceAllChildren(content);
            }
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props, children, mappedChildren) {
        return el(
            "div",
            {
                class: classNames({
                    [ NAMESPACE ]: true,
                    [ props.bodyBgStyle ]: props.bodyBgStyle,
                    [ props.bodyClassNames ]: props.bodyClassNames
                })
            },
            this.renderChildren()
        );
    }
}
